---
title: "Analysis 3: Introgression Analysis"
author: "Joe Gunn"
date: "2022-07-28"
output: html_document
---

# Project: Interspecific Hybridization in the Smallmouth Bass species complex : Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Spotted Bass (SPB; <i>Micropterus punctulatus</i>), Smallmouth Bass (SMB; <i>M. dolomieu</i>), the newly elevated Neosho Bass (NB; <i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin, the Ouachita Bass (OB; <i>M. cf. dolomieu </i> Ouachita River), and the Little River Bass (LRB; <i>M. cf. dolomieu </i> Little River).</font>

## Specific Aim: Hierarchical structure analysis with diagnostic markers
For this aim, we are using regression-based genomic cline analysis in the R package Introgress to assess the degree of introgression within black bass species in the Central Interior Highlands (CIH), with emphasis on populations in the Arkansas River Basin, Ouachita River Basin, and Little River Basin. We obtained SNP marker data from from SPB, SMB, and NB, and two potentially distinct species in the Ouachita River Basin, OB and LRB, which were known to be diagnostic of these species (Long et al. 2020). Specifically, we aim to determine the relative timing of introgression in backcrosses and F2 hybrids by regressing hybrid index (estimate of shared genomic ancestry between parental populations) on interspecific heterozygosity (freqency of heterozygous genotypes derived from different parental alleles).

## Phases of Analysis
### Phase 2: Introgression analysis

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(adegenet)
library(genetics) # this package is needed for internal functions within the code for Introgress (see source code below)
library(hybriddetective) # had to be installed with package "remote"

### Read in source code for pacakge "Introgress"
source("code/introgress_source_code/R/calc.intersp.het.R")
source("code/introgress_source_code/R/prepare.data.R")
source("code/introgress_source_code/R/clines.plot.R")
source("code/introgress_source_code/R/compare.clines.R")
source("code/introgress_source_code/R/delta.R")
source("code/introgress_source_code/R/est.h.R")
source("code/introgress_source_code/R/fit.c.clines.R")
source("code/introgress_source_code/R/fit.invariant.clines.R")
source("code/introgress_source_code/R/fixup.combos.touse.R")
source("code/introgress_source_code/R/genomic.clines.R")
source("code/introgress_source_code/R/h.func.R")
source("code/introgress_source_code/R/like.h.R")
source("code/introgress_source_code/R/mk.image.R")
source("code/introgress_source_code/R/per.locus.like.R")
source("code/introgress_source_code/R/prepare.data.R")
source("code/introgress_source_code/R/s.wrapper.R")
source("code/introgress_source_code/R/support.limit.R")
source("code/introgress_source_code/R/test.combinations.R")
source("code/introgress_source_code/R/test.data.objects.R")
source("code/introgress_source_code/R/test.genotypes.R")
source("code/introgress_source_code/R/triangle.plot.R")
```

## PHASE 1: INTROGRESSION ANALYSIS
In this phase of the analysis, we read in raw SNP genotype data along with hybrid category inference from NewHybrids analysis (see `snolh_structure_analysis.Rmd`, Phase 2 - 4, Step 3) for each hierarchical analysis conducted in Structure: 1) SPB vs. the SBSC); 2) SMB vs. all other species in the CIH; 3) all species within the CIH

Programs need:

Introgress v.1.2 (Gompert and Buerkle 2010)

Citation:

Gompert Z, Buerkle CA. 2010. Introgress: A software package for mapping components of isolation in hybrids. Molecular Ecology Resources 10: 378-384. doi: 10.1111/j.1755-0998.2009.02733.x

### STEP 1: Read in and prepare posterior probability data for each hybrid analysis in the program NewHybrids (see `snolh_structure_analysis.Rmd`, Phase 1, Steps 2 - 4).
In this step, we read in the posterior probability output data from NewHybrids analysis for each hierarchical dataset (SPB vs. the SMBC; SMB vs. CIH speceis; CIH species), which we conducted in Analysis 2: Structure and Hybrid Assignment Analysis. Specifically, we add labels to each output dataset for "hybrid assignment" (one of six hybrid categories) based on posterior probability greater than 0.5. 

#### 1a: Load the fully filtered dataset ('full_data'); run the Rmd chunk below.

##### Load in full, filtered data:
```{r}
load("../filtering_analysis/data/processed_raw_data/full_data.rda")
```

#### 1b. Read in and prepare data for hybridization between SPB and SMBC (`snolh_structure_analysis.Rmd`, Phase 1, Step 2).

##### 1b.1. Load in data; run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/spb_smbc_output/spb_smbc_pofz.rda")
```

##### 1b.2. Gather, clean, and prepare data for introgress analysis; run the Rmd chunk below.

##### Prepare data:
```{r}
# Get pure SMBC individuals 
spb_smbc_pure_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get pure SPB individuals
spb_smbc_pure_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F1 individuals
spb_smbc_f1_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F2 individuals
spb_smbc_f2_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get SMBC backcross individuals
spb_smbc_bc_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get SPB backcross individuals
spb_smbc_bc_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Generate list of all individuals by hybrid category
hybrid_metadata <- rbind(spb_smbc_pure_smbc_metadata, 
                           spb_smbc_pure_spb_metadata, 
                           spb_smbc_f1_metadata, 
                           spb_smbc_f2_metadata, 
                           spb_smbc_bc_smbc_metadata, 
                           spb_smbc_bc_spb_metadata)

# Gather full data so that metadata and genotype data can be merged in subsequent step
full_data <- full_data %>%
  dplyr::select(Structure_Num:OUOU_locus14857_59)

# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
hybrid_data <- merge(hybrid_metadata, full_data, by = "Structure_Num") 

# Select only SPB other SNPs
hybrid_gdata <- hybrid_data %>%
  dplyr::select(matches("SM_SP"))

# Create dataframe to hold locus information, which is needed for introgress analysis. First column is the locus name, and the second column is a "type" description of the type of marker, in our case we arbitrarily give the designation "c" for "codominant"
loci_data <- as.data.frame(colnames(hybrid_gdata[,1:15]))
loci_data <- cbind(loci_data, rep("c", times = 15))
colnames(loci_data) <- c("locus","type")

# Extract vectors with names of individuals (Structure_Num) and hybrid category
indivs <- as.vector(hybrid_data[,1])
hybrid_category <- factor(hybrid_data[,2])

# Convert data to genind object
genind <- df2genind(hybrid_gdata, 
                    ncode = 3, 
                    ind.names = indivs, 
                    pop = hybrid_category, 
                    ploidy = 2)

# Convert from genind back to dataframe (this is necessary to code the alleles properly with a "/")
hybrid_gdata <- genind2df(genind, oneColPerAll = FALSE, sep="/")

# Create data object for parent 1 genotypes (SPB)
p1_data <- hybrid_gdata %>%
  filter(pop == "pure_spb") %>%
  dplyr::select(-c(pop))

# Create data object for parent 2 genotypes (IH)
p2_data <- hybrid_gdata %>%
  filter(pop == "pure_ih") %>%
  dplyr::select(-c(pop))

# Create data object for "admixed" genotypes (all other individuals)
admix_data <- hybrid_gdata %>%
  filter(pop != "pure_ih") %>%
  filter(pop != "pure_spb") %>%
  dplyr::select(-c(pop))

# Transpose each object created above for introgress input format
p1_data <- t(p1_data)
p2_data <- t(p2_data)
admix_data <- t(admix_data)
```

##### 1b.3. Run introgress analysis; run the Rmd chunk below.

##### Run introgress:
```{r}
# Generate introgress input formatted data
introgress_input <- prepare.data(admix.gen = admix_data, 
                                 loci.data = loci_data, 
                                 parental1 = p1_data, 
                                 parental2 = p2_data, 
                                 pop.id = F, 
                                 ind.id = F, 
                                 fixed = FALSE, 
                                 sep.rows = FALSE, 
                                 sep.columns = FALSE)

# Calculate hybrid index for all individuals
h_index <- est.h(introgress.data = introgress_input, 
                 loci.data = loci_data)

# Calculate interspecific heterozygosity for all individuals
int_het <- calc.intersp.het(introgress.data = introgress_input)

# Create dataframe for interspecific heterozygosity values
int_het <- int_het %>%
  as.data.frame()

colnames(int_het) <- "int_het"

# Get hybrid metadata for admixed samples
hybrids <- hybrid_data %>%
  filter(hybrid_category != "pure_spb") %>%
  filter(hybrid_category != "pure_smbc") %>%
  dplyr::select(hybrid_category)

# Create full dataset with hybrid metadata, hybrid index, and interspecific heterozygosity
introgress_data <- cbind(hybrids, h_index, int_het)
```
##### 1b.4. Plot introgress results using a triangle plot; run the Rmd chunk below.

##### Generate triangle plot for SPB and the SMBC: `02_spb_smbc_introgress.pdf`
```{r}
# Generate triangle plot, regressing interspecific heterozygosity on hybrid index
pdf("figures/02_spb_smbc_introgress.pdf", width=8, height=6)

ggplot(introgress_data, aes(x = h, y = int_het, fill = hybrid_category)) + 
  geom_errorbar(aes(xmin=lower, xmax=upper), width = .1) +
  geom_point(alpha = 0.8, 
             color = "black", 
             pch = 21, 
             size = 4) +
  
  theme_set(theme_cowplot(12)) + 
  xlim(0,1) +
  ylim(0,1.5) +
  scale_fill_manual(values = c("#56106EFF","#F98C0AFF")) + 
  labs(x = "Hybrid index", y = "Interspecific heterozygosity") + 
  theme(axis.title = element_text(size = 20, color = "black")) + 
  theme(axis.text = element_text(size = 15, color = "black")) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_line(color = "black")) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

#### 1c. Read in and prepare data for hybridization between SMB and CIH species (`snolh_structure_analysis.Rmd`, Phase 1, Step 3).

##### 1c.1. Load in data; run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/smb_cih_output/smb_cih_pofz.rda")
```

##### 1c.2. Gather, clean, and prepare data for introgress analysis; run the Rmd chunk below.

##### Prepare data:
```{r}
# Get pure CIH individuals 
smb_cih_pure_cih_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get pure SPB individuals
smb_cih_pure_smb_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F1 individuals
smb_cih_f1_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F2 individuals
smb_cih_f2_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get IH backcross individuals
smb_cih_bc_cih_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get SPB backcross individuals
smb_cih_bc_smb_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Generate list of all individuals by hybrid category
hybrid_metadata <- rbind(smb_cih_pure_cih_metadata, 
                           smb_cih_pure_smb_metadata, 
                           smb_cih_f1_metadata, 
                           smb_cih_f2_metadata, 
                           smb_cih_bc_cih_metadata, 
                           smb_cih_bc_smb_metadata)

# Gather full data so that metadata and genotype data can be merged in subsequent step
full_data <- full_data %>%
  dplyr::select(Structure_Num:OUOU_locus14857_59)

# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
hybrid_data <- merge(hybrid_metadata, full_data, by = "Structure_Num") 

# Select only SPB other SNPs
hybrid_gdata <- hybrid_data %>%
  dplyr::select(matches("N_SM"))

# Create dataframe to hold locus information, which is needed for introgress analysis. First column is the locus name, and the second column is a "type" description of the type of marker, in our case we arbitrarily give the designation "c" for "codominant"
loci_data <- as.data.frame(colnames(hybrid_gdata[,1:33]))
loci_data <- cbind(loci_data, rep("c", times = 33))
colnames(loci_data) <- c("locus","type")

# Extract vectors with names of individuals (Structure_Num) and hybrid category
indivs <- as.vector(hybrid_data[,1])
hybrid_category <- factor(hybrid_data[,2])

# Convert data to genind object
genind <- df2genind(hybrid_gdata, 
                    ncode = 3, 
                    ind.names = indivs, 
                    pop = hybrid_category, 
                    ploidy = 2)

# Convert from genind back to dataframe (this is necessary to code the alleles properly with a "/")
hybrid_gdata <- genind2df(genind, oneColPerAll = FALSE, sep="/")

# Create data object for parent 1 genotypes (SPB)
p1_data <- hybrid_gdata %>%
  filter(pop == "pure_smb") %>%
  dplyr::select(-c(pop))

# Create data object for parent 2 genotypes (IH)
p2_data <- hybrid_gdata %>%
  filter(pop == "pure_cih") %>%
  dplyr::select(-c(pop))

# Create data object for "admixed" genotypes (all other individuals)
admix_data <- hybrid_gdata %>%
  filter(pop != "pure_cih") %>%
  filter(pop != "pure_smb") %>%
  dplyr::select(-c(pop))

# Transpose each object created above for introgress input format
p1_data <- t(p1_data)
p2_data <- t(p2_data)
admix_data <- t(admix_data)
```

##### 1c.3. Run introgress analysis; run the Rmd chunk below.

##### Run introgress:
```{r}
# Generate introgress input formatted data
introgress_input <- prepare.data(admix.gen = admix_data, 
                                 loci.data = loci_data, 
                                 parental1 = p1_data, 
                                 parental2 = p2_data, 
                                 pop.id = F, 
                                 ind.id = F, 
                                 fixed = FALSE, 
                                 sep.rows = FALSE, 
                                 sep.columns = FALSE)

# Calculate hybrid index for all individuals
h_index <- est.h(introgress.data = introgress_input, 
                 loci.data = loci_data)

# Calculate interspecific heterozygosity for all individuals
int_het <- calc.intersp.het(introgress.data = introgress_input)

# Create dataframe for interspecific heterozygosity values
int_het <- int_het %>%
  as.data.frame()

colnames(int_het) <- "int_het"

# Get hybrid metadata for admixed samples
hybrids <- hybrid_data %>%
  filter(hybrid_category != "pure_smb") %>%
  filter(hybrid_category != "pure_cih") %>%
  dplyr::select(hybrid_category)

# Create full dataset with hybrid metadata, hybrid index, and interspecific heterozygosity
introgress_data <- cbind(hybrids, h_index, int_het)
```

##### 1c.4. Plot introgress results using a triangle plot; run the Rmd chunk below.

##### Generate triangle plot for SMB and species in the CIH: `03_smb_cih_introgress.pdf`
```{r}
# Generate triangle plot, regressing interspecific heterozygosity on hybrid index
pdf("figures/03_smb_cih_introgress.pdf", width=8, height=6)

ggplot(introgress_data, aes(x = h, y = int_het, fill = hybrid_category)) + 
  geom_errorbar(aes(xmin=lower, xmax=upper), width = .1) +
  geom_point(alpha = 0.8, 
             color = "black", 
             pch = 21, 
             size = 4) +
  theme_set(theme_cowplot(12)) + 
  xlim(0,1) +
  ylim(0,1.5) +
  scale_fill_manual(values = c("#56106EFF","#F98C0AFF")) + 
  labs(x = "Hybrid index", y = "Interspecific heterozygosity") + 
  theme(axis.title = element_text(size = 20, color = "black")) + 
  theme(axis.text = element_text(size = 15, color = "black")) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_line(color = "black")) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

#### 1d. Read in and prepare data for hybridization among species in the CIH (`snolh_structure_analysis.Rmd`, Phase 1, Step 4).
We found no hybrid individuals when we tested for hybridization between NB and LRB or between NB and OB. However, we found one hybrid when we assessed hybridization between LRB and OB; introgression analysis for this individual is below.

##### 1d.3. Read in and prepare data for hybridization between LRB and OB; run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/lrb_ob_output/lrb_ob_pofz.rda")
```


```{r}
load("../structure_analysis/data/newhybrids_data/output_data/spb_smbc_output/spb_smbc_pofz.rda")

# Get pure SMBC individuals 
spb_smbc_pure_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get pure SPB individuals
spb_smbc_pure_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F1 individuals
spb_smbc_f1_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get F2 individuals
spb_smbc_f2_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get SMBC backcross individuals
spb_smbc_bc_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Get SPB backcross individuals
spb_smbc_bc_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(Structure_Num, hybrid_category)

# Generate list of all individuals by hybrid category
hybrid_metadata <- rbind(spb_smbc_pure_smbc_metadata, 
                           spb_smbc_pure_spb_metadata, 
                           spb_smbc_f1_metadata, 
                           spb_smbc_f2_metadata, 
                           spb_smbc_bc_smbc_metadata, 
                           spb_smbc_bc_spb_metadata)

# Gather full data so that metadata and genotype data can be merged in subsequent step
full_data <- full_data %>%
  dplyr::select(Structure_Num:OUOU_locus14857_59)

# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
hybrid_data <- merge(hybrid_metadata, full_data, by = "Structure_Num") 

# Select only SPB other SNPs
hybrid_data <- hybrid_data %>%
  dplyr::select(Structure_Num, hybrid_category, matches("SM_SP"))

hybrid_data <- with(hybrid_gdata, hybrid_gdata[order(hybrid_category),])

spb_smbc_pures <- hybrid_data %>%
  filter(hybrid_category == "pure_spb" | hybrid_category == "pure_smbc") 

spb_smbc_hybrids <- hybrid_data %>%
  filter(hybrid_category == "bc_spb" | hybrid_category == "bc_smbc", hybrid_category == "f1", hybrid_category == "f2")

write_xlsx(spb_smbc_pures, path = "data/hybrid_simulation_data/spb_smbc_pures.xlsx")
write_xlsx(spb_smbc_hybrids, path = "data/hybrid_simulation_data/spb_smbc_hybrids.xlsx")

pop_rename_smbc_spb <- data.frame(Opop = c(rep("", times = 446)), Rename = c(rep("smbc", times = 441), rep("spb", times = 5)))

#Designate the two different populations (Pure Neosho, Pure Northern)
subset_genepop_rename("data/hybrid_simulation_data/spb_smbc_pure_genepop.txt", pop_rename_smbc_spb, path = "data/hybrid_simulation_data/spb_smbc_pure_genepop_edited.txt")



genepop_allelefreq("data/hybrid_simulation_data/spb_smbc_pure_genepop_edited.txt")

freqbasedsim_GTFreq("data/hybrid_simulation_data/spb_smbc_pure_genepop_edited.txt")

```

