---
title: "Analysis 2 - Phase 3: Hybridization between Smallmouth Bass and Central Interior Highlands Species"
author: "Joe Gunn"
date: "2022-08-13"
output: html_document
---

# Project: Central Interior Highlands Hybridization: Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the black basses (genus Micropterus) in the Central Interior Highlands (CIH) ecoregion, including natural and anthropogenic hybridization between Spotted Bass (SPB; <i>Micropterus punctulatus</i>) and species within the Smallmouth Bass species complex (SMBC): Smallmouth Bass (SMB; <i>M. dolomieu</i>), the newly elevated Neosho Bass (NB; <i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin, the Ouachita Bass (OB; <i>M. cf. dolomieu </i> Ouachita River), and the Little River Bass (LRB; <i>M. cf. dolomieu </i> Little River).</font>

## Specific Aim: Hierarchical structure and hybrid detection in Smallmouth Bass and Central Interior Highlands Species
For this aim, we are using Bayesian inference with STRUCTURE and NEWHYBRIDS to assess patterns and levels of hybridization between SMB, NB, OB, and LRB.

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(writexl)
library(genepopedit)
library(pophelper)
library(parallelnewhybrid)
```

### STEP 1: Data preparation

#### 1a: Load in fully filtered dataset ('full_data') and list of hybrids detected between Spotted Bass and all other Central Interior Highlands fish ('spb_other_hybrids'); run the Rmd chunk below to load in all data.

##### Load in full, filtered data and list of SPB hybrids:
```{r}
load("../../filtering_analysis/data/processed_raw_data/full_data.rda")
load("../data/filtering_data/spb_smbc_hybrids.rda")
```

#### 1b: Filter metadata and SNP genotype data 
In this step, we are filtering the metadata and SNP genotype data to only include samples of interest for analysis of population structure and hybridization between SMB and all other CIH species. For this analysis, these data will include SMB, NB, OB, and LRB, excluding SPB hybrids detected in Analysis 2, Phase 03, and only SNP loci that were found to be diagnostic between SMB and all other CIH species (SMB-SNPS)

##### 1b.1. Filter metadata; run the Rmd chunk below:

##### Filter metadata to only include SMB and CIH species, excluding hybrids with SPB
```{r}
## get metadata for c-binding later
smb_cih_metadata <- full_data[1:13]

## Filter out SPB from the metadata
smb_cih_metadata <- smb_cih_metadata %>%
  filter(putative_taxon != "SPB") 

# Remove hybrids detected with SPB
smb_cih_metadata <- smb_cih_metadata[ ! smb_cih_metadata$structure_number %in% spb_smbc_hybrids, ] 

## Convert factors to characters
smb_cih_metadata <- smb_cih_metadata %>%
  mutate(structure_number = as.character(structure_number),
         river = as.character(river))
```

##### 1.b.2. Filter SNP genotype data; run the Rmd chunk below:

##### Filter out genotype data into diagnostic SNP groups for SMB and CIH species
```{r}
## Filter out SPB from genotype data
smb_cih_snps <- full_data %>%
  filter(putative_taxon != "SPB")

# Remove hybrids detected with Spotted Bass (Analysis 2, Step 1b (1b_spb_other.Rmd))
smb_cih_snps <- smb_cih_snps[ ! smb_cih_snps$structure_number %in% spb_smbc_hybrids, ]

# Get only SMB-SNPS loci (N_SM)
smb_cih_snps <- smb_cih_snps %>%
  dplyr::select(matches("N_SM"))
```

<b>Details on SMB and CIH species dataset:</b> <br>
<i>N</i><sub>samples</sub> = 441 <br>
<i>N</i><sub>loci</sub> = 33 <br>

#### 1c: Combine metadata and snp datasets for combinations of interest (see above); Run the Rmd chunk below to combine datasets.

##### Combine metadata and snp datasets and generate excel/Rda files: files are stored in the directory `data/processed_genotype_data/`
```{r}
# cbind metadata and genotype data
smb_cih <- cbind(smb_cih_metadata, smb_cih_snps)

## Generate excel files to hold genotype data
write_xlsx(smb_cih, path = "../data/processed_genotype_data/smb_cih.xlsx")
```

#### 1d: Generate genepop file format in Excel.
In this step, we manually converted the Excel file generated in step 1b to basic Genepop format, carefully following the steps below:

##### 1d.1. Remove all metadata columns from the dataset except for "river", which designates the a priori stream population for a given sample, and "structure_number", which gives a structure-formatted individual label for each sample

##### 1d.2. Add a "," after each structure_number

##### 1d.3. Insert a row above each new population in the data, and insert "pop" in the structure_number column in each empty row (the word "pop" should appear for each population)

##### 1d.4. Delete the "river" column

##### 1d.5. Copy the header row of SNP IDs and transpose paste in a new Excel sheet; each SNP ID should be listed in its own row (number of rows should be equal to numbrer of SNP IDs)

##### 1d.6. Copy all genotype and label data from the original Excel file and paste in the row immediately below the list of SNP IDs

##### 1d.7. Save the Excel sheet as a .xlsx and a .txt file

#### 1e: Generate STRUCTURE formatted files from genepop format
In this step, we are reading in, checking, and converting genepop files for combined datasets into STRUCTURE formatted files. There are quite a few tricky and very specific formatting needs for the genepop text file before genepopedit will convert successfully. Follow Step 3a below closely to create a proper genepop format. Use Ctrl + F to find and replace the necessary things.

##### IMPORTANT NOTE ON DATA: genepopedit requires that sample names be in the format 'popname_01', 'popname_02', etc. where the 'popname' is a signifier for the population of origin for the sample, and the '01', etc. are numerical designations for the samples. 'popname' and the numeric must be separated by a "_", and there can only be a single "_". Otherwise, genepopedit will not read the sample's population of origin correctly. In the case of this project, we are clumping samples together by stream for STRUCTURE analysis to see if there may be substructure mapped by stream. Many sample IDs in our dataset were not in the required format, so we added an additional column to the metadata set and associated processed data sets called "structure_ID", which gives a unique sample name in the proper structure format. This column can then be linked back to any other column in downstream analyses if we need to know which specific samples the results came from.

##### 1e.1. Follow the guide for the library 'genepopedit' (Stanley et al. [year]) to prepare genepop files before converting them to STRUCTURE format.

###### 1e.1.1 All "," in the genepop text files were manually replaced with " ,  "

###### 1e.1.2. All "tabs" between six-digit alleles were were manually replaced with " " (a single space).

###### 1e.1.3. The top row designating the dataset combination was manually omitted.

###### 1e.1.4. Any additional space at the bottom of the text file was removed.

##### 1e.2. Generate STRUCTURE files from genepop files.
Here it is important to not include the 'popgroup = ' unless it is explicitly necessary. Assigning popgroup to a dataframe with population names will cause the STRUCTURE input file to have word strings in the popdata column. Our experience is that STRUCTURE expects an integer here rather than a string, so best not to use popgroup.

##### Convert genepop to STRUCTURE format: files are stored in the directory `data/structure_data/`
```{r}
# Generate structure file for SMB and CIH species
genepop_structure("../data/genepop_data/smb_cih_genepop.txt", 
                  locusnames = TRUE, 
                  #popgroup = "data/genepop_data/smb_cih_pops.csv",
                  path = "../data/structure_data/input_data/smb_cih_structure.txt")
```

<b>Details on SMB and CIH species:</b> <br>
<i>N</i><sub>populations</sub> = 23

#### 1f: Generate batch list files and shell scripts for running STRUCTURE in parallel
In this step, we are creating batch lists of command line code to run STRUCTURE analyses in parallel. We are making a single batch list for each of our data combinations (generated above; 'spb_other', 'smb_cih', 'ih', 'ouachita_little', 'ouachita', 'little'). Each batch list will contain a separate line of code to run a single replicate at an a priori determined number of populations (<i>K</i>, listed in the chunk above for each dataset). We will run each analysis in 10 replicates at each <i>K</i>, e.g.:

structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_1
structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_2
structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_3
structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_4
structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_5
.
.
.
structure -K 1 -m mainparams -e extraparams -i structure_input.txt -o structure_output_1_10
structure -K 2 -m mainparams -e extraparams -i structure_input.txt -o structure_output_2_1
structure -K 2 -m mainparams -e extraparams -i structure_input.txt -o structure_output_2_2
.
.
.
where, in the output name, the first number represents the K value for the run, and the last number represents the replicate.

##### 1f.1. Generate batch list file; run the Rmd chunk below.
The Rmd chunk below generates a list of commands that contains the full cluster paths (/home/data/...) for necessary files.

##### Generate batch commands: `batch_cmd_lists/smb_cih_structure_batch_cmd_list.txt`
```{r}
# N(pops) = 23

nk <- data.frame(c(1:23))
nreps <- data.frame(c(1:10))

cat("", file="batch_cmd_lists/smb_cih_structure_batch_cmd_list.txt")

# Run loop to create file for storing commands
for(ii in 1:nrow(nk)) {

  for(aa in 1:nrow(nreps)) {
    
    structure_call <- paste("structure ")

    param_files <- paste(" -m /home/jcg5g9/data/SNOLH_Genomics/structure_analysis/data/structure_data/param_files/mainparams_smb_cih -e /home/jcg5g9/data/SNOLH_Genomics/structure_analysis/data/structure_data/param_files/extraparams")

    input <- paste(" -i /home/jcg5g9/data/SNOLH_Genomics/structure_analysis/data/structure_data/input_data/smb_cih_structure.txt")
  
    output <- paste(" -o /home/jcg5g9/data/SNOLH_Genomics/structure_analysis/data/structure_data/output_data/smb_cih_structure_output/")
  
    cat(paste(structure_call, "-K ", nk[ii,], param_files, input, output,
            strsplit("smb_cih_structure.txt", "_structure.txt", fixed = TRUE)[[1]][1], "_", nk[ii,], "_", nreps[aa,], sep=""),
      "\n", 
      file = paste("batch_cmd_lists/smb_cih_structure_batch_cmd_list.txt"), # name file to store list
      append = TRUE)
  }
}
```

##### 1f.2. Generate shell script; run the Rmd chunk below.

##### IMPORTANT NOTE: This shell script was generated on the server directly, and thus the Rmd chunk below does not need to be run to generate the file. This is purely to keep a record of each script file. 

##### Generate batch commands: `shell_scripts/smb_cih_structure.sh`
```{r}
#!/bin/bash
#-------------------------------------------------------------------------------
#  SBATCH CONFIG
#-------------------------------------------------------------------------------
## resources
#SBATCH --partition Lewis
#SBATCH --nodes=1
#SBATCH --ntasks=1  # used for MP#SBATCH -e error_%A_%a.err # Standard errorI codes, otherwise leav$
##SBATCH --cpus-per-task=12  # cores per task
#SBATCH --mem-per-cpu=16G  # memory per core (default is 1GB/core)
#SBATCH --time 2-00:00  # days-hours:minutes
#SBATCH --qos=normal
#SBATCH --array=23-220

## labels and outputs
#SBATCH --job-name=snolh_structure_jgunn
#
#SBATCH -o test_%A_%a.out # Standard output
#SBATCH -e error_%A_%a.err # Standard error

## notifications
#SBATCH --mail-user=jcg5g9@mail.missouri.edu  # email address for notifications
#SBATCH --mail-type=END,FAIL  # which type of notifications to send
#-------------------------------------------------------------------------------

#echo "### Starting at: $(date) ###"


# load packages
#module load rss/rss-2020
#module load structure/structure-2.3.4

#COMMANDA=`head -n ${SLURM_ARRAY_TASK_ID} ../batch_cmd_lists/smb_cih_structure_batch_cmd_list.txt | tail -n 1`
#eval $COMMANDA


#echo "### Ending at: $(date) ###"
```

#### 1g: Prepare mainparam and extraparam files for STRUCTURE.
In this step, we are preparing the mainparam and extraparam input files for STRUCTURE so that they are unique to each analysis we are running.

##### 1g.1. Edit base mainparam STRUCTURE file (downloaded with STRUCTURE program) and generate a separate, unique mainparam file.

<b>mainparam file:</b> <br>
`../data/structure_data/param_files/mainparams_smb_cih` <br> 

###### 1g.1.1 Edit "maxpops" value to reflect the number of populations designated in Step 3b above. These values should be the following:

<b>Populationss:</b> 23 <br>

###### 1g.1.2. Set the number of burn-in and MCMC iterations to run; these are the same for each analysis:

<b>Burn-in runs:</b> 500,000 <br>
<b>MCMC runs:</b> 1,000,000 <br>

###### 1g.1.3. Set the number of individuals. These values should be the following:

<b>Individuals:</b> 441 <br>

###### 1g.1.4. Set the number of loci. These values should be the following:

<b>Loci:</b> 33 <br>

###### 1g.1.5. Set 'ONEROWPERIND' to '0'

###### 1g.1.6. Set 'LABEL' to '1'

###### 1g.1.7. Set 'POPDATA' to '1'

###### 1g.1.8. Set 'POPFLAG' to '0'

##### 1g.2. Edit base extraparam STRUCTURE file (downloaded with STRUCTURE program).
We did not change any of the default settings in the extraparams file (most importantly, we used the default admixture model), so we only used a single extraparams file.

<b>extraparam file:</b> <br>
`../data/structure_data/param_files/extraparams` <br> 

### STEP 2: population Structure Analysis

#### 2a: Run STRUCTURE analysis using the input data generated in STEP 1. Navigate to `shell_scripts/` Be sure that all relative and full paths to all input files and output destination directories are set up properly (ideally, this is already done within this GitHub repo). This command line code assumes capability to run the code using SLURM or a SLURM-like cluster scheduling software.

Run `sbatch smb_cih_structure.sh`

#### 2b: Structure output files are generated and stored here: `../data/structure_data/output_data/`.

#### 2c: Compress output directory into a zip file compatible with Structure Selector (Li and Liu 2017) or Structure Harvester (Earl and vonHoldt 2011) online.

#### 2d: Submit zip directory to Structure Selector or Structure Harvester to extract summary results.

#### 2e: Copy detlaK table (Evanno et al. 2005) and save as excel file here: `../data/structure_data/summary_data`
We further analyze deltak results graphically in Step 2g.4.

#### 2f: Copy puechmaille table (Puechmaille et al. 2016) and save as excel file here:
`../data/structure_data/summary_data`
We found that Puechmaille metrics universally support K=4, and we graph the results for deltak and Puechmaille in Step 2g.5.

#### 2g: Visualize STRUCTURE runs for all data combinations.

##### 2g.1. Convert STRUCTURE files into aligned Q files compatible with analysis in the program CLUMPP (Jakobbson and Rosenberg 2007); run the rmd chunk below.

##### Convert STRUCTURE files to aligned Q files for CLUMPP and generate deltaK table: 
```{r}
# Get a list of structure files for each data combination
smb_cih_sfiles <- list.files("../data/structure_data/output_data/smb_cih_structure_output/", full.names = T)

# Extract q value information (ancestry proportions) from each run for each K value for each individual from the STRUCTURE output files in the directories listed above
smb_cih_Q <- readQ(smb_cih_sfiles)

# Tabulate information from the q lists
smb_cih_tab <- tabulateQ(smb_cih_Q)

# Summarize information from tabultions above
smb_cih_summary <- summariseQ(smb_cih_tab)

# Extract deltaK and associated summary information using Evanno method
smb_cih_evanno <- evannoMethodStructure(smb_cih_summary, returnplot = F) 

# Set infinity and NA to zero arbitrarily
smb_cih_evanno$deltaK[smb_cih_evanno$deltaK == "Inf"] <- 0
smb_cih_evanno$deltaK[is.na(smb_cih_evanno$deltaK)] <- 0

# Write Evanno table to Excel table for manuscript preparation.

## Convert to data frame
smb_cih_evanno <- as.data.frame(smb_cih_evanno)

## Write Excel file
write_xlsx(smb_cih_evanno, "../data/structure_data/deltak_data/smb_cih_deltak.xlsx")

# Align replicate runs for each K to correct label switching
smb_cih_align <- alignK(smb_cih_Q)
```

##### 2g.2. Export CLUMPP compatbile files for CLUMPP analysis; run the rmd chunk below to export CLUMPP associated files for each K.

<b>We used the following parameters: </b> <br>

Large-K-Greedy algorithm (paramrep = 3) <br>
10000 replicates <br>

### IMPORTANT NOTE: This step only needs to be run ONCE to generate files for CLUMPP. Once you have run this chunk, move on to Step 2g.3. Uncomment each line to run this code.

##### Export CLUMPP files:
```{r}
clumppExport(smb_cih_align, 
            parammode = 3, 
            paramrep = 10000, 
            exportpath = "../data/structure_data/clumpp_data/smb_cih_clumpp")
```

The code above generates a .txt file (pop_K#-combined.txt) with combined cluster ancestry proportions for each individual at each K and stores it in a separate directory labeled "pop_K#", where '#' is the corresponding K value. Additionally, the code generates an accompanying "paramfile" that is input for CLUMPP, which contains information on parameters to generate merged datasets across our 10 replicates at each K.

##### 2g.3. Generate merged Q files using CLUMPP.
In this step, we used the software program CLUMPP (Jakobbson and Rosenberg 2007) to merge cluster ancestry proportion inferences for all replicate runs of each K across individuals. We downloaded the Linux 64-bit build of CLUMPP v.1.1.2 and installed it on our computing cluster. We then executed the program for each analysis by sequentially copying the associated paramfiles generated in the clumppExport function (see section above) to the CLUMPP home directory. All output files were then moved to the corresponding structure directory.

###### 2g.3.1. Downolad CLUMPP and install on cluster; navigate to the desired directory: `code`. Run the command: `wget https://rosenberglab.stanford.edu/software/CLUMPP_Linux64.1.1.2.tar.gz`

###### 2g.3.2. Run the command: `gunzip CLUMPP_Linux64.1.1.2.tar.gz`

###### 2g.3.3. Run the command: `tar -xvf CLUMPP_Linux64.1.1.2.tar`

###### 2g.3.4. For each analysis separately, copy the 'pop_K#-combined.txt' file and 'paramfile' over to the CLUMPP directory (`CLUMPP_Linux64.1.1.2`)

###### 2g.3.5. For each analysis separately, execute CLUMPP on the paramfile; run the code `./CLUMPP paramfile`
We are running this code with GREEDY OPTION 2, which uses random sampling of replicate runs. 

This code generates three additional files, which we moved back to the corresponding pop_K folder in the clumpp_data directory:

<b>Files generated: </b> <br>

pop_K#-combined-miscfile.txt <br>
pop_K#-combined-merged.txt <br>
pop_K#-combined-aligned.txt <br>

##### 2g.4. Plot deltaK results; run the rmd chunk below.

##### deltak results for SMB and other Central Interior Highlands populations: `03_smb_cih_deltaK.pdf` 
```{r}
# DeltaK results
pdf("../figures/structure_figures/deltak_plots/03_smb_cih_deltaK.pdf", width = 8, height = 5)

ggplot(smb_cih_evanno, aes(x = k, y = deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 3, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(smb_cih_evanno$k), breaks = smb_cih_evanno$k)

dev.off()
```

We found strongest support for K=3, but also support for K=4 using the deltak metric. We found universal support for K=4 using the Puechmaille metric. We therefore present result for K=3 and K=4 in Step 2g.5 below. Results for deltaK and Puechmaille metrics are given in Tables S3 and S4, respectively.

##### 2g.5. Plot STRUCTURE runs at best K value (K=3 and K=4); run the rmd chunk below.

##### STRUCTURE results of SMB and all other Central Interior Highlands populations; 1) `03_smb_cih_k3.pdf` and 2) `03_smb_cih_k4.pdf`
```{r}
# Read in combined-merged Q table from CLUMPP
smb_cih_k3 <- readQ("../data/structure_data/clumpp_data/smb_cih_clumpp/pop_K3/pop_K3-combined-merged.txt")

# Visualize Structure
plotQ(smb_cih_k3,
      grplab = smb_cih[,c(13,7)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "river",
      subsetgrp = c("Lake_Superior_Chequamegon_Bay", "Lake_Erie_Central_Basin_Ohio", "Oneida_Lake_NY", "Smith_Fork_Creek", "Skiatook_Lake", "Honey_Creek", "Spavinaw_Creek", "Baron_Fork", "Caney_Creek", "Lee_Creek", "Blackfork_Creek", "Honobia_Creek", "Little_River", "Glover_River", "Big_Eagle_Creek", "Upper_Mountain_Fork", "Mountain_Fork", "Pero_Creek_Rolling_Fork_trib", "Cossatot_River_AR", "Western_Saline_River_AR", "Little_Missouri_River_AR", "Caddo_River_AR", "Ouachita_River_AR", "Eastern_Saline_River_AR"),
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.5,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("#1B9E77","#7570B3","#E7298A"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "03_smb_cih_k3",
      exportpath = "../figures/structure_figures/q_plots")

# Read in combined-merged Q table from CLUMPP
smb_cih_k4 <- readQ("../data/structure_data/clumpp_data/smb_cih_clumpp/pop_K4/pop_K4-combined-merged.txt")

# Visualize Structure
plotQ(smb_cih_k4,
      grplab = smb_cih[,c(13,7)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "river",
      subsetgrp = c("Lake_Superior_Chequamegon_Bay", "Lake_Erie_Central_Basin_Ohio", "Oneida_Lake_NY", "Smith_Fork_Creek", "Skiatook_Lake", "Honey_Creek", "Spavinaw_Creek", "Baron_Fork", "Caney_Creek", "Lee_Creek", "Blackfork_Creek", "Honobia_Creek", "Little_River", "Glover_River", "Big_Eagle_Creek", "Upper_Mountain_Fork", "Mountain_Fork", "Pero_Creek_Rolling_Fork_trib", "Cossatot_River_AR", "Western_Saline_River_AR", "Little_Missouri_River_AR", "Caddo_River_AR", "Ouachita_River_AR", "Eastern_Saline_River_AR"),
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.5,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("#1B9E77","#7570B3","#E7298A","#E6AB02"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "03_smb_cih_k4",
      exportpath = "../figures/structure_figures/q_plots")

# Read in combined-merged Q table from CLUMPP
smb_cih_k2 <- readQ("../data/structure_data/clumpp_data/smb_cih_clumpp/pop_K2/pop_K2-combined-merged.txt")

# Visualize Structure
plotQ(smb_cih_k2,
      grplab = smb_cih[,c(13,7)],
      showticks = F,
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "river",
      subsetgrp = c("Lake_Superior_Chequamegon_Bay", "Lake_Erie_Central_Basin_Ohio", "Oneida_Lake_NY", "Smith_Fork_Creek", "Skiatook_Lake", "Honey_Creek", "Spavinaw_Creek", "Baron_Fork", "Caney_Creek", "Lee_Creek", "Blackfork_Creek", "Honobia_Creek", "Little_River", "Glover_River", "Big_Eagle_Creek", "Upper_Mountain_Fork", "Mountain_Fork", "Pero_Creek_Rolling_Fork_trib", "Cossatot_River_AR", "Western_Saline_River_AR", "Little_Missouri_River_AR", "Caddo_River_AR", "Ouachita_River_AR", "Eastern_Saline_River_AR"),
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.5,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("#1B9E77","#C7EAE5"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "03_smb_cih_k2",
      exportpath = "../figures/structure_figures/q_plots")
```

These figures are the bases for Figures S3a, S3b, and Figures S2b 3b, respectively.


#### 2h: Analyze ancestry proportions from Smallmouth Bass for all river populations in the CIH; run the Rmd cbhunk below:
In this step, we summarize ancestry proportions for each river calculated in STRUCTURE. Specifically, we calculate the average SMB ancestry across rivers.

##### Summarize ancestry proportions for rivers:
```{r}
# column bind metadata with q values determined in STRUCTURE
smb_cih_k3 <- cbind(smb_cih_metadata, smb_cih_k3)

# select only the river column, along with the columns containing q values from structure
smb_cih_k3 <- smb_cih_k3[,c(7,14:16)]

# rename columns to be more descriptive
colnames(smb_cih_k3) <- c("river", "smb_ancestry", "cih_ancestry_1", "cih_ancestry_2")

# Omit SPB individuals (first five rows of dataset)
smb_cih_k3 <- smb_cih_k3[-c(419:441),]

# Summarize average ancestry of SPB per river
smb_ancestry_summary_k3 <- smb_cih_k3 %>%
  group_by(river) %>%
  summarize(mean_smb_k3 = mean(SMB_ancestry))

# arrange in descending order
smb_ancestry_summary_k3 %>% 
  arrange(desc(mean_smb_k3))

# Given that there are no major clustering patterns in ancestry within the CIH species (non-SMB) at K=3, we also summarize ancestry proportions of SMB at K=2 to see if they are equal. If they are equal, we will use results for K=2 to describe ancestry patterns

# column bind metadata with q values determined in STRUCTURE
smb_cih_k2 <- cbind(smb_cih_metadata, smb_cih_k2)

# select only the river column, along with the columns containing q values from structure
smb_cih_k2 <- smb_cih_k2[,c(7,14:15)]

# rename columns to be more descriptive
colnames(smb_cih_k2) <- c("river", "smb_ancestry", "cih_ancestry")

# Omit SPB individuals (first five rows of dataset)
smb_cih_k2 <- smb_cih_k2[-c(419:441),]

# Summarize average ancestry of SPB per river
smb_ancestry_summary_k2 <- smb_cih_k2 %>%
  group_by(river) %>%
  summarize(mean_smb_k2 = mean(SMB_ancestry))

# arrange in descending order
smb_ancestry_summary_k2 %>% 
  arrange(desc(mean_smb_k2))

# compare ancestry calculations for K=2 and K=3 back to back
smb_ancestry <- cbind(smb_ancestry_summary_k3, smb_ancestry_summary_k2[,2])

# calculate difference between ancestry calculations for K=2 and K=3
options(scipen = 999)
smb_ancestry <- smb_ancestry %>%
  mutate(mean_smb_diff = mean_smb_k3 - mean_smb_k2)

# View distribution of ancestry calculation difference values
ggplot(smb_ancestry, aes(mean_smb_diff)) + 
  geom_histogram(binwidth = 0.0009)
```

### STEP 3: Hybridization analysis with NEWHYBRIDS
In this step, we are running complementary analyses on the dataset with SMB and all other samples in the software program NewHybrids, which uses Bayesian inference of genotype data to assign individuals to one of six <i>a priori</i> modeled hybrid classes (parent 1, parent 2, F1 hybrid, F2 hybrid, backcross to parent 1, or backcross to parent 2).

See `snolh_structure_analysis.Rmd` (Line 190 - 201) for programs needed for analysis.

#### 3a: Convert genepop file to NEWHYBRIDS format and generate 'individual file' for analysis; run the Rmd chunk below:

Running the Rmd chunk below generates two files: 
    1) input for NewHybrids (`smb_cih_nh.txt`),
    2) accompanying individual file (`smb_cih_individuals.txt`)
    
These files are stored here: `../data/newhybrids_data/input_data/`

##### Generate NEWHYBRIDS input file and accompanying individual file:
```{r}
# Generate newhybrids file for SMB and CIH species
genepop_newhybrids("../data/genepop_data/smb_cih_genepop.txt", 
                   path = "../data/newhybrids_data/input_data/smb_cih_nh.txt")

# Generate individual file (list of sample names ('structure_number') to accompany NewHybrids)
smb_cih_individuals <- data.frame(indiv = as.character(smb_cih_metadata$structure_number))

# Write individual file to .txt format
write_tsv(smb_cih_individuals, "../data/newhybrids_data/input_data/smb_cih_individuals.txt")
```

#### 3b: Install newhybrids on local machine (or plan to use a remote server, setting up paths as needed)
NewHybrids was installed on a local machine here: `/Users/joegunn/Downloads/newhybrids/`

##### 3b.1. Set the number of burn-in and MCMC iterations to run:

<b>Burn-in runs:</b> 100,000 <br>
<b>MCMC runs:</b> 500,000 <br>

#### 3c: Run parallelnewhybrids for MAC; run the Rmd chunk below to run the parallel newhybrids program on the input data.

##### Run NewHybrids program with 100,000 burn-in and 500,000 MCMC sweeps
```{r}
parallelnh_OSX("../data/newhybrids_data/input_data/", 
               where.NH = "/Users/joegunn/Downloads/newhybrids/", 
               burnin = 100000, 
               sweeps = 500000)
```

The parallelnewhybrids program will generate results and store them in a new directory (`NH.Results`) within the same directory as the input data. Since we are running multiple sequential analyses, we chose to rename the resulting output directory and store the input data within that directory along with the associated results so as not to confuse which input files go with which analyses. See specific steps below.

##### 3c.1. Manually rename output directory: `smb_cih_output/`

##### 3c.2. Manually move all output and input files into the newly named directory.

##### 3c.3. Move the output folder into the `output_data/` directory within `newhybrids_data/`.

#### 3d: Visualize NEWHYBRIDS results; run the Rmd chunk below to read in, clean, and plot NewHybrids output.

##### Visualize NEWHYBRIDS output and generate plot for SMB and all other Central Interior Highlands species: `smb_cih.pdf`
```{r}
# Read in NewHybrids "PofZ" output txt file
smb_cih_pofz <- read_tsv("../data/newhybrids_data/output_data/smb_cih_output/smb_cih_nh.txt_PofZ.txt")

# Omit first two columns (the first column is individual number and the second is empty)
smb_cih_pofz <- data.frame(smb_cih_pofz[,-c(1:2)])

# Change column names to indicate hybrid class clearly
colnames(smb_cih_pofz) <- c("pure_cih","pure_smb","f1","f2","bc_cih","bc_smb")

# Bind metadata to NewHybrids output to match posterior probabilities with individual samples
smb_cih_pofz <- cbind(smb_cih_metadata, smb_cih_pofz)

# Gather dataset for plotting and convert variables to factors/characters. 'structure_number' was converted to a character in this step, and then factor levels were set in the next step to ensure that individuals were graphed in the same order as in STRUCTURE (for later direct one-to-one comparison)
smb_cih_pofz <- smb_cih_pofz %>%
  gather(c(14:19), key = "hybrid_category", value = "probability") %>%
  mutate(hybrid_category = factor(hybrid_category)) %>%
  mutate(structure_number = as.character(structure_number))

# Change levels to get the correct order (see explanation for step above)
smb_cih_pofz$river <- factor(smb_cih_pofz$river, levels = c("Lake_Superior_Chequamegon_Bay", "Lake_Erie_Central_Basin_Ohio", "Oneida_Lake_NY", "Smith_Fork_Creek", "Skiatook_Lake", "Honey_Creek", "Spavinaw_Creek", "Baron_Fork", "Caney_Creek", "Lee_Creek", "Blackfork_Creek", "Honobia_Creek", "Little_River", "Glover_River", "Big_Eagle_Creek", "Upper_Mountain_Fork", "Mountain_Fork", "Pero_Creek_Rolling_Fork_trib", "Cossatot_River_AR", "Western_Saline_River_AR", "Little_Missouri_River_AR", "Caddo_River_AR", "Ouachita_River_AR", "Eastern_Saline_River_AR"))

# Convert  analysis ordercolumn to factor
smb_cih_pofz <- smb_cih_pofz %>%
  mutate(analysis_order = factor(analysis_order))

# Reorder analysis order to be in the same order as Rivers (same as presented in Structure q plot)
smb_cih_pofz <- smb_cih_pofz %>% 
  mutate(analysis_order = fct_reorder(analysis_order, as.integer(river)))
 

# Save dataframe for future use in introgress analysis (Analysis 3)
save(smb_cih_pofz, file = "../data/newhybrids_data/output_data/smb_cih_output/smb_cih_pofz.rda")

# Plot posterior probability of assignment to each hybrid category as a cumulative bar plot for each individual (colors represent assignment probabilities for each of six possible hybrid categories)
pdf("../figures/newhybrids_figures/03_smb_cih.pdf", width = 120, height = 12)

ggplot(smb_cih_pofz, aes(x = analysis_order, y = probability, fill = hybrid_category)) + 
   geom_bar(stat = "identity", show.legend = T, color = "black") +
   geom_hline(yintercept = 0.5, color = "red", linetype = "longdash", size = 3) +
   theme_set(theme_cowplot(12)) +
   labs(x = "Individual", y = "Cumulative probability of hybrid ID") +
   scale_fill_manual(values = c("blue","red","green","yellow","white","white")) +
   theme(axis.title.x = element_blank())  +
   theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
   theme(axis.text = element_text(angle = 90)) +
   scale_y_continuous(expand=c(0,0))

dev.off()
```

This figure is the basis for Figure S2b in the manuscript.

#### 3e: Identify individuals of hybrid status (non-pure, F1, F2, or backcross generation hybrids) across the whole dataset (these individuals are removed in subsequent hierarchical analyses)
Before moving on to assess population structure and hybridization among CIH species (NB, OB, and LRB), we are removing any hybrids resulting from crossing with SMB.

We are designating individuals as hybrid status if the posterior probability of assignment to any of the four hybrid categories (F1, F2, Back-cross to Parent 1, Backross to Parent 2) is greater than 0.50. We are basing this threshold on Long et al. (2021), who performed an in-depth power analysis following the hybriddetective workflow (Wringe et al. 2017) for each hybrid category at various numbers of SNP loci for varying posterior probability cutoffs. They found that the highest power to accurately detect hybrid individuals (and avoid type 2 error, failure to reject the null hypothesis when it is false) occurs at a probability cutoff of 0.5.

##### 3e.1. Generate a list of hybrid individual sample names to be removed from subsequent datasets and save as an .Rda.

##### Generate and save list of hybrid individuals
```{r}
# Summarize all hybrids together by population
smb_cih_hybrid_summary <- smb_cih_pofz %>%
  filter(probability > 0.5) %>%
  group_by(river, hybrid_category) %>%
  count()

# Get pure CIH individuals 
smb_cih_pure_cih <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 396 pure CIH individuals

# Get pure SMB individuals
smb_cih_pure_smb <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 23 pure SMB

# Get F1 individuals
smb_cih_f1 <- smb_cih_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 0 F1 hybrids

# Get F2 individuals
smb_cih_f2 <- smb_cih_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 4 F2 hybrids

# Gets CIH backcross individuals
smb_cih_bc_ih <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 18 CIH backcrosses

# Get pSMB backcross individuals
smb_cih_bc_smb <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number)

### 0 SMB backcrosses

# Generate list of "pure" samples (CIH and SMB)
smb_cih_pure <- rbind(smb_cih_pure_cih, 
                      smb_cih_pure_smb)

# Generate list of "hybrid" samples (f1, f2, and backcrosses)
smb_cih_hybrids <- rbind(smb_cih_f1,
                         smb_cih_f2, 
                         smb_cih_bc_ih, 
                         smb_cih_bc_smb)

# Select the first column and convert to data frame
smb_cih_hybrids <- data.frame(smb_cih_hybrids)

# Save the list of bad samples as a .txt file without column names
write_tsv(smb_cih_hybrids, 
          file = "../data/filtering_data/smb_cih_hybrids.txt", 
          col_names = FALSE)

# Convert hybrids dataframe to vector for saving and loading later
smb_cih_hybrids <- as.vector(smb_cih_hybrids[,1])

# Save list to be loaded in 03_snolh_structure_ih.Rmd
save(smb_cih_hybrids, file = "../data/filtering_data/smb_cih_hybrids.rda")
```

## ------------------------ END OF PHASE 3 OF STRUCTURE AND HYBRID ASSIGNMENT ANALYSIS ----------------------- ##
