---
title: "Analysis 3: Introgression Analysis"
author: "Joe Gunn"
date: "2022-07-28"
output: html_document
---

# Project: Interspecific Hybridization in the Smallmouth Bass species complex : Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Spotted Bass (SPB; <i>Micropterus punctulatus</i>), Smallmouth Bass (SMB; <i>M. dolomieu</i>), the newly elevated Neosho Bass (NB; <i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin, the Ouachita Bass (OB; <i>M. cf. dolomieu </i> Ouachita River), and the Little River Bass (LRB; <i>M. cf. dolomieu </i> Little River).</font>

## Specific Aim: Hierarchical structure analysis with diagnostic markers
For this aim, we are using regression-based genomic cline analysis in the R package Introgress to assess the degree of introgression within black bass species in the Central Interior Highlands (CIH), with emphasis on populations in the Arkansas River Basin, Ouachita River Basin, and Little River Basin. We obtained SNP marker data from from SPB, SMB, and NB, and two potentially distinct species in the Ouachita River Basin, OB and LRB, which were known to be diagnostic of these species (Long et al. 2020). Specifically, we aim to determine the relative timing of introgression in backcrosses and F2 hybrids by regressing hybrid index (estimate of shared genomic ancestry between parental populations) on interspecific heterozygosity (frequency of heterozygous genotypes derived from different parental alleles).

## Phases of Analysis
### Phase 1: Introgression analysis

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(adegenet)
library(genetics) # this package is needed for internal functions within Introgress (see source code below)
library(hybriddetective) # We installed a forked version of the hybriddetective package from GitHub (stevemussmann/hybriddetective), which fixed a code issue that was generating an error with converting GenePop files to vector objects. See https://github.com/bwringe/hybriddetective/issues/12

### Read in source code for pacakge "Introgress"
source("code/introgress_source_code/R/calc.intersp.het.R")
source("code/introgress_source_code/R/prepare.data.R")
source("code/introgress_source_code/R/clines.plot.R")
source("code/introgress_source_code/R/compare.clines.R")
source("code/introgress_source_code/R/delta.R")
source("code/introgress_source_code/R/est.h.R")
source("code/introgress_source_code/R/fit.c.clines.R")
source("code/introgress_source_code/R/fit.invariant.clines.R")
source("code/introgress_source_code/R/fixup.combos.touse.R")
source("code/introgress_source_code/R/genomic.clines.R")
source("code/introgress_source_code/R/h.func.R")
source("code/introgress_source_code/R/like.h.R")
source("code/introgress_source_code/R/mk.image.R")
source("code/introgress_source_code/R/per.locus.like.R")
source("code/introgress_source_code/R/prepare.data.R")
source("code/introgress_source_code/R/s.wrapper.R")
source("code/introgress_source_code/R/support.limit.R")
source("code/introgress_source_code/R/test.combinations.R")
source("code/introgress_source_code/R/test.data.objects.R")
source("code/introgress_source_code/R/test.genotypes.R")
source("code/introgress_source_code/R/triangle.plot.R")
```

## PHASE 1: INTROGRESSION ANALYSIS
In this phase of the analysis, we read in raw SNP genotype data along with hybrid category inference from NewHybrids analysis (see `snolh_structure_analysis.Rmd`, Phase 2 - 4, Step 3) for each hierarchical analysis conducted in Structure: 1) SPB vs. the SBSC); 2) SMB vs. all other species in the CIH; 3) all species within the CIH

Programs need:

Introgress v.1.2 (Gompert and Buerkle 2010)

Citation:

Gompert Z, Buerkle CA. 2010. Introgress: A software package for mapping components of isolation in hybrids. Molecular Ecology Resources 10: 378-384. doi: 10.1111/j.1755-0998.2009.02733.x

### STEP 1: Load the fully filtered dataset ('full_data'); run the Rmd chunk below.
In this step, we load in the fully filtered dataset, including metadata and genotype data, which will be merged with hybrid category inference output from NewHybrids analysis in the steps below.

##### Load in full, filtered data:
```{r}
## Load in full data
load("../filtering_analysis/data/processed_raw_data/full_data.rda")

# Gather full data so that metadata and genotype data can be merged in subsequent step
full_data <- full_data %>%
  dplyr::select(structure_number:OUOU_locus14857_59)
```

### STEP 2: Conduct introgression analysis for hybridization between SPB and SMBC.
In this step, we assess the relationship between hybrid index and interspecific heterozygosity to infer introgression between SPB and the SMBC.

#### 2a: Load in posterior probability output data from NewHybrids analysis of hybridization between SPB and SMBC (`snolh_structure_analysis.Rmd`, Phase 2); run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/spb_smbc_output/spb_smbc_pofz.rda")
```

#### 2b: Gather, clean, and merge hybrid metadata with full data for SPB and SMBC; run the Rmd chunk below.

##### Prepare hybrid metadata:
```{r}
# Get pure SMBC individuals 
spb_smbc_pure_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get pure SPB individuals
spb_smbc_pure_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "pure_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number,
                putative_taxon,
                hybrid_category)

# Get F1 individuals
spb_smbc_f1_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get F2 individuals
spb_smbc_f2_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get SMBC backcross individuals
spb_smbc_bc_smbc_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_smbc") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get SPB backcross individuals
spb_smbc_bc_spb_metadata <- spb_smbc_pofz %>%
  filter(hybrid_category == "bc_spb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Generate list of all individuals by hybrid category
spb_smbc_hybrid_metadata <- rbind(spb_smbc_pure_smbc_metadata, 
                                  spb_smbc_pure_spb_metadata, 
                                  spb_smbc_f1_metadata, 
                                  spb_smbc_f2_metadata, 
                                  spb_smbc_bc_smbc_metadata, 
                                  spb_smbc_bc_spb_metadata)

# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
spb_smbc_hybrid_data <- merge(spb_smbc_hybrid_metadata, 
                              full_data, 
                              by = "structure_number") 

# Select only SPB-SNPs
spb_smbc_hybrid_genotypes <- spb_smbc_hybrid_data %>%
  dplyr::select(matches("SM_SP"))
```

#### 2c: Prepare hybrid data for introgression analysis; run the Rmd chunk below.

##### Prepare hybrid data:
```{r}
# Create dataframe to hold locus information, which is needed for introgress analysis. First column is the locus name, and the second column is a "type" description of the type of marker, in our case we arbitrarily give the designation "c" for "codominant"
spb_smbc_loci_data <- as.data.frame(colnames(spb_smbc_hybrid_genotypes[,1:15]))
spb_smbc_loci_data <- cbind(spb_smbc_loci_data, rep("c", times = 15))
colnames(spb_smbc_loci_data) <- c("locus","type")

# Extract vectors with names of individuals (structure_number) and hybrid category
spb_smbc_indivs <- as.vector(spb_smbc_hybrid_data[,1])
spb_smbc_hybrid_category <- factor(spb_smbc_hybrid_data[,3])

# Convert data to genind object
spb_smbc_genind <- df2genind(spb_smbc_hybrid_genotypes, 
                             ncode = 3, 
                             ind.names = spb_smbc_indivs,
                             pop = spb_smbc_hybrid_category, 
                             ploidy = 2)

# Convert from genind back to dataframe (this is necessary to code the alleles properly with a "/")
spb_smbc_hybrid_genotypes <- genind2df(spb_smbc_genind, oneColPerAll = FALSE, sep="/")

# Create data object for parent 1 genotypes (SPB)
spb_p1_data <- spb_smbc_hybrid_genotypes %>%
  filter(pop == "pure_spb") %>%
  dplyr::select(-c(pop))

# Create data object for parent 2 genotypes (IH)
smbc_p2_data <- spb_smbc_hybrid_genotypes %>%
  filter(pop == "pure_smbc") %>%
  dplyr::select(-c(pop))

# Create data object for "admixed" genotypes (all other individuals)
spb_smbc_admix_data <- spb_smbc_hybrid_genotypes %>%
  filter(pop != "pure_smbc") %>%
  filter(pop != "pure_spb") %>%
  dplyr::select(-c(pop))

# Transpose each object created above for introgress input format
spb_p1_data <- t(spb_p1_data)
smbc_p2_data <- t(smbc_p2_data)
spb_smbc_admix_data <- t(spb_smbc_admix_data)
```

#### 2d: Run introgress analysis; run the Rmd chunk below.

##### Run introgression analysis:
```{r}
# Generate introgress input formatted data
spb_smbc_introgress <- prepare.data(admix.gen = spb_smbc_admix_data, 
                                    loci.data = spb_smbc_loci_data, 
                                    parental1 = spb_p1_data, 
                                    parental2 = smbc_p2_data, 
                                    pop.id = F, 
                                    ind.id = F, 
                                    fixed = FALSE, 
                                    sep.rows = FALSE, 
                                    sep.columns = FALSE)

# Calculate hybrid index for all individuals
spb_smbc_hi <- est.h(introgress.data = spb_smbc_introgress, 
                 loci.data = spb_smbc_loci_data)

# Calculate interspecific heterozygosity for all individuals
spb_smbc_inth <- calc.intersp.het(introgress.data = spb_smbc_introgress)

# Create dataframe for interspecific heterozygosity values
spb_smbc_inth <- spb_smbc_inth %>%
  as.data.frame()
colnames(spb_smbc_inth) <- "int_het"

# Get hybrid metadata for admixed samples
spb_smbc_hybrids <- spb_smbc_hybrid_data %>%
  filter(hybrid_category != "pure_spb") %>%
  filter(hybrid_category != "pure_smbc") %>%
  dplyr::select(hybrid_category)

# Create full dataset with hybrid metadata, hybrid index, and interspecific heterozygosity
spb_smbc_introgress_data <- cbind(spb_smbc_hybrids, 
                                  spb_smbc_hi, 
                                  spb_smbc_inth)

# Get metadata on putative taxon to append to the introgress dataset (putative taxon will be in the same order as the introgress dataframe)
spb_smbc_species <- spb_smbc_hybrid_data %>%
  filter(hybrid_category != "pure_smbc") %>%
  filter(hybrid_category != "pure_spb") %>%
  dplyr::select(putative_taxon)

# Bind metadata and introgress data
spb_smbc_introgress_data <- cbind(spb_smbc_species, 
                                  spb_smbc_introgress_data)

# Arrange data in order of hybrid index
spb_smbc_introgress_data %>% arrange(h)
```

#### 2e: Plot introgression analysis results using a triangle plot (relationship between hybrid index and interspecific heterozygosity); run the Rmd chunk below.

##### Generate triangle plot for SPB and the SMBC: `02_spb_smbc_introgress.pdf`
```{r}
# Generate triangle plot, regressing interspecific heterozygosity on hybrid index
pdf("figures/02_spb_smbc_introgress.pdf", width = 5, height = 4)

ggplot(spb_smbc_introgress_data, aes(x = h, y = int_het)) + 
  geom_errorbar(aes(xmin=lower, xmax=upper, color = putative_taxon),
                width = .1, 
                stat = "identity",
                show.legend = F) +
  geom_point(aes(fill = putative_taxon,
             shape = hybrid_category),
             alpha = 0.8, 
             size = 4, 
             color = "black",
             stat = "identity",
             show.legend = F,
             position = position_jitter()) +
  theme_set(theme_cowplot(12)) + 
  xlim(0,1) +
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01)) +
  ylim(0,1.2) +
  scale_color_manual(values = c("#E6AB02","#7570B3","#E7298A")) + 
  scale_fill_manual(values = c("#E6AB02","#7570B3","#E7298A")) + 
  scale_shape_manual(values=c(21,22)) +
  labs(x = "Hybrid index", y = "Interspecific heterozygosity", fill = "Species", color = "Species", shape = "Hybrid Category") + 
  theme(legend.position = c(0.03, 0.15)) +
  theme(legend.background = element_rect(color = "black", size = 0.75)) + 
  theme(legend.margin = margin(5, 5, 5, 5)) +
  theme(legend.title = element_text(size = 20, face = "bold")) +
  theme(legend.text = element_text(size = 20)) +
  theme(axis.title = element_text(size = 20, color = "black")) + 
  theme(axis.text = element_text(size = 15, color = "black")) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_line(color = "black")) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

This figure is the basis for Figure 4a in the manuscript.

### STEP 3: Conduct introgression analysis for hybridization between SMB and CIH.
In this step, we assess the relationship between hybrid index and interspecific heterozygosity to infer introgression between SMB and species in the CIH.

#### 3a: Load in posterior probability output data from NewHybrids analysis of hybridization between SMB and species in the CIH (`snolh_structure_analysis.Rmd`, Phase 3); run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/smb_cih_output/smb_cih_pofz.rda")
```

#### 3b. Gather, clean, and merge hybrid metadata with full data for SMB and the CIH; run the Rmd chunk below.

##### Prepare hybrid metadata:
```{r}
# Get pure SMB individuals
smb_cih_pure_smb_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get pure CIH individuals 
smb_cih_pure_cih_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "pure_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon,
                hybrid_category)

# Get F1 individuals
smb_cih_f1_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get F2 individuals
smb_cih_f2_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number,
                putative_taxon, 
                hybrid_category)

# Get CIH backcross individuals
smb_cih_bc_cih_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_cih") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Get SMB backcross individuals
smb_cih_bc_smb_metadata <- smb_cih_pofz %>%
  filter(hybrid_category == "bc_smb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                putative_taxon, 
                hybrid_category)

# Generate list of all individuals by hybrid category
smb_cih_hybrid_metadata <- rbind(smb_cih_pure_smb_metadata,
                                 smb_cih_pure_cih_metadata, 
                                 smb_cih_f1_metadata, 
                                 smb_cih_f2_metadata, 
                                 smb_cih_bc_cih_metadata, 
                                 smb_cih_bc_smb_metadata)

# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
smb_cih_hybrid_data <- merge(smb_cih_hybrid_metadata, 
                             full_data, 
                             by = "structure_number") 

# Select only SMB-SNPs
smb_cih_hybrid_genotypes <- smb_cih_hybrid_data %>%
  dplyr::select(matches("N_SM"))
```

#### 3c: Prepare hybrid data for introgression analysis; run the Rmd chunk below.

##### Prepare hybrid data:
```{r}
# Create dataframe to hold locus information, which is needed for introgress analysis. First column is the locus name, and the second column is a "type" description of the type of marker, in our case we arbitrarily give the designation "c" for "codominant"
smb_cih_loci_data <- as.data.frame(colnames(smb_cih_hybrid_genotypes[,1:33]))
smb_cih_loci_data <- cbind(smb_cih_loci_data, rep("c", times = 33))
colnames(smb_cih_loci_data) <- c("locus","type")

# Extract vectors with names of individuals (structure_number) and hybrid category
smb_cih_indivs <- as.vector(smb_cih_hybrid_data[,1])
smb_cih_hybrid_category <- factor(smb_cih_hybrid_data[,3])

# Convert data to genind object
smb_cih_genind <- df2genind(smb_cih_hybrid_genotypes, 
                            ncode = 3, 
                            ind.names = smb_cih_indivs, 
                            pop = smb_cih_hybrid_category, 
                            ploidy = 2)

# Convert from genind back to dataframe (this is necessary to code the alleles properly with a "/")
smb_cih_hybrid_genotypes <- genind2df(smb_cih_genind, oneColPerAll = FALSE, sep="/")

# Create data object for parent 1 genotypes (SPB)
smb_cih_p1_data <- smb_cih_hybrid_genotypes %>%
  filter(pop == "pure_smb") %>%
  dplyr::select(-c(pop))

# Create data object for parent 2 genotypes (IH)
smb_cih_p2_data <- smb_cih_hybrid_genotypes %>%
  filter(pop == "pure_cih") %>%
  dplyr::select(-c(pop))

# Create data object for "admixed" genotypes (all other individuals)
smb_cih_admix_data <- smb_cih_hybrid_genotypes %>%
  filter(pop != "pure_cih") %>%
  filter(pop != "pure_smb") %>%
  dplyr::select(-c(pop))

# Transpose each object created above for introgress input format
smb_cih_p1_data <- t(smb_cih_p1_data)
smb_cih_p2_data <- t(smb_cih_p2_data)
smb_cih_admix_data <- t(smb_cih_admix_data)
```

#### 3d: Run introgress analysis; run the Rmd chunk below.

##### Run introgression analysis:
```{r}
# Generate introgress input formatted data
smb_cih_introgress <- prepare.data(admix.gen = smb_cih_admix_data, 
                                   loci.data = smb_cih_loci_data, 
                                   parental1 = smb_cih_p1_data,
                                   parental2 = smb_cih_p2_data, 
                                   pop.id = F, 
                                   ind.id = F, 
                                   fixed = FALSE, 
                                   sep.rows = FALSE, 
                                   sep.columns = FALSE)

# Calculate hybrid index for all individuals
smb_cih_hi <- est.h(introgress.data = smb_cih_introgress, 
                 loci.data = smb_cih_loci_data)

# Calculate interspecific heterozygosity for all individuals
smb_cih_inth <- calc.intersp.het(introgress.data = smb_cih_introgress)

# Create dataframe for interspecific heterozygosity values
smb_cih_inth <- smb_cih_inth %>%
  as.data.frame()

# Rename column
colnames(smb_cih_inth) <- "int_het"

# Get hybrid metadata for admixed samples
smb_cih_hybrids <- smb_cih_hybrid_data %>%
  filter(hybrid_category != "pure_smb") %>%
  filter(hybrid_category != "pure_cih") %>%
  dplyr::select(hybrid_category)

# Create full dataset with hybrid metadata, hybrid index, and interspecific heterozygosity
smb_cih_introgress_data <- cbind(smb_cih_hybrids, 
                                 smb_cih_hi, 
                                 smb_cih_inth)

# Get metadata on putative taxon to append to the introgress dataset (putative taxon will be in the same order as the introgress dataframe)
smb_cih_species <- smb_cih_hybrid_data %>%
  filter(hybrid_category != "pure_smb") %>%
  filter(hybrid_category != "pure_cih") %>%
  dplyr::select(putative_taxon)

# Bind metadata and introgress data
smb_cih_introgress_data <- cbind(smb_cih_species, 
                                 smb_cih_introgress_data)

# Arrange data in order of hybrid index
smb_cih_introgress_data %>% arrange(h)
```

#### 2e: Plot introgression analysis results using a triangle plot (relationship between hybrid index and interspecific heterozygosity); run the Rmd chunk below.

##### Generate triangle plot for SMB and the CIH: `03_smb_cih_introgress.pdf`
```{r}
# Generate triangle plot, regressing interspecific heterozygosity on hybrid index
pdf("figures/03_smb_cih_introgress.pdf", width = 5, height = 4)

ggplot(smb_cih_introgress_data, aes(x = h, y = int_het)) + 
    geom_errorbar(aes(xmin = lower, xmax = upper, color = putative_taxon),
                width = .1, 
                stat = "identity",
                show.legend = F) +
  geom_point(aes(fill = putative_taxon,
             shape = hybrid_category),
             alpha = 0.8, 
             size = 4, 
             color = "black",
             stat = "identity",
             show.legend = F,
             position = position_jitter()) +
  theme_set(theme_cowplot(12)) + 
  xlim(0,1) +
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01)) +
  ylim(0,1.2) +
  scale_color_manual(values = c("#E6AB02","#7570B3","#E7298A")) + 
  scale_fill_manual(values = c("#E6AB02","#7570B3","#E7298A")) + 
  scale_shape_manual(values=c(21,22)) +
  labs(x = "Hybrid index", y = "Interspecific heterozygosity", fill = "Species", color = "Species", shape = "Hybrid Category") + 
  theme(legend.position = c(0.03, 0.15)) +
  theme(legend.background = element_rect(color = "black", size = 0.75)) + 
  theme(legend.margin = margin(5, 5, 5, 5)) +
  theme(legend.title = element_text(size = 20, face = "bold")) +
  theme(legend.text = element_text(size = 20)) +
  theme(axis.title = element_text(size = 20, color = "black")) + 
  theme(axis.text = element_text(size = 15, color = "black")) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_line(color = "black")) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

This figure is the basis for Figure 4b in the manuscript.

### STEP 4: Conduct introgression analysis for hybridization between LRB and OB
In this step, we assess the relationship between hybrid index and interspecific heterozygosity to infer introgression between LRB and OB, which is the only species pair that had an inferred hybrid in NewHybries analysis 

#### 4a: Load in posterior probability output data from NewHybrids analysis of hybridization between LRB and OB (`snolh_structure_analysis.Rmd`, Phase 4); run the Rmd chunk below.

##### Load in saved .Rda file
```{r}
load("../structure_analysis/data/newhybrids_data/output_data/lrb_ob_output/lrb_ob_pofz.rda")
```

#### 4b. Gather, clean, and merge hybrid metadata with full data for LRB and OB; run the Rmd chunk below.

##### Prepare hybrid metadata:
```{r}
# Get pure SMB individuals
lrb_ob_pure_lrb_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "pure_lrb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get pure CIH individuals 
lrb_ob_pure_ob_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "pure_ob") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get F1 individuals
lrb_ob_f1_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "f1") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get F2 individuals
lrb_ob_f2_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "f2") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get CIH backcross individuals
lrb_ob_bc_lrb_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "bc_lrb") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Get SMB backcross individuals
lrb_ob_bc_ob_metadata <- lrb_ob_pofz %>%
  filter(hybrid_category == "bc_ob") %>%
  filter(probability > 0.5) %>%
  dplyr::select(structure_number, 
                hybrid_category)

# Generate list of all individuals by hybrid category
lrb_ob_hybrid_metadata <- rbind(lrb_ob_pure_lrb_metadata,
                                lrb_ob_pure_ob_metadata, 
                                lrb_ob_f1_metadata, 
                                lrb_ob_f2_metadata, 
                                lrb_ob_bc_lrb_metadata, 
                                lrb_ob_bc_ob_metadata)


# Merge hybrid metadata (hybrid category inferred in NewHybrids) with genotype data
lrb_ob_hybrid_data <- merge(lrb_ob_hybrid_metadata, 
                            full_data, 
                            by = "structure_number") 

# Select only SMB-SNPs
lrb_ob_hybrid_genotypes <- lrb_ob_hybrid_data %>%
  dplyr::select(matches("NEO"),
         matches("OUOU"),
         matches("OULR"))
```

#### 4c: Prepare hybrid data for introgression analysis; run the Rmd chunk below.

##### Prepare hybrid data:
```{r}
# Create dataframe to hold locus information, which is needed for introgress analysis. First column is the locus name, and the second column is a "type" description of the type of marker, in our case we arbitrarily give the designation "c" for "codominant"
lrb_ob_loci_data <- as.data.frame(colnames(lrb_ob_hybrid_genotypes[,1:138]))
lrb_ob_loci_data <- cbind(lrb_ob_loci_data, rep("c", times = 138))
colnames(lrb_ob_loci_data) <- c("locus","type")

# Extract vectors with names of individuals (structure_number) and hybrid category
lrb_ob_indivs <- as.vector(lrb_ob_hybrid_data[,1])
lrb_ob_hybrid_category <- factor(lrb_ob_hybrid_data[,2])

# Convert data to genind object
lrb_ob_genind <- df2genind(lrb_ob_hybrid_genotypes, 
                            ncode = 3, 
                            ind.names = lrb_ob_indivs, 
                            pop = lrb_ob_hybrid_category, 
                            ploidy = 2)

# Convert from genind back to dataframe (this is necessary to code the alleles properly with a "/")
lrb_ob_hybrid_genotypes <- genind2df(lrb_ob_genind, oneColPerAll = FALSE, sep = "/")

# Create data object for parent 1 genotypes (SPB)
lrb_ob_p1_data <- lrb_ob_hybrid_genotypes %>%
  filter(pop == "pure_lrb") %>%
  dplyr::select(-c(pop))

# Create data object for parent 2 genotypes (IH)
lrb_ob_p2_data <- lrb_ob_hybrid_genotypes %>%
  filter(pop == "pure_ob") %>%
  dplyr::select(-c(pop))

# Create data object for "admixed" genotypes (all other individuals)
lrb_ob_admix_data <- lrb_ob_hybrid_genotypes %>%
  filter(pop != "pure_lrb") %>%
  filter(pop != "pure_ob") %>%
  dplyr::select(-c(pop))

# Transpose each object created above for introgress input format
lrb_ob_p1_data <- t(lrb_ob_p1_data)
lrb_ob_p2_data <- t(lrb_ob_p2_data)
lrb_ob_admix_data <- t(lrb_ob_admix_data)
```

#### 4d: Run introgress analysis; run the Rmd chunk below.

##### Run introgression analysis:
```{r}
# Generate introgress input formatted data
lrb_ob_introgress <- prepare.data(admix.gen = lrb_ob_admix_data, 
                                  loci.data = lrb_ob_loci_data, 
                                  parental1 = lrb_ob_p1_data,
                                  parental2 = lrb_ob_p2_data, 
                                  pop.id = F, 
                                  ind.id = F, 
                                  fixed = FALSE, 
                                  sep.rows = FALSE, 
                                  sep.columns = FALSE)
```

The above code for generating introgress input data did not work for only one hybrid individual (given that the code is copied and modified from the working code above for SPB vs. SMBC and SMB vs. CIH, I presume that the failure to run this function is due to sample size and not due to coding error). Thus, we do not continue on with this analysis for the single OB-LRB back-cross individual. 

## ------------------------ END OF PHASE 1: INTROGRESSION ANALYSIS ----------------------- ##

## ------------------------ END OF ANALYSIS 3: INTROGRESSION ANALYSIS ----------------------- ##
