---
title: "Analysis 1: Filtering Analysis"
author: "Joe Gunn"
date: "2022-07-28"
output: html_document
---

# Project: Interspecific Hybridization in the Smallmouth Bass species complex : Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Spotted Bass (SPB; <i>Micropterus punctulatus</i>), Smallmouth Bass (SMB; <i>M. dolomieu</i>), the newly elevated Neosho Bass (NB; <i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin, the Ouachita Bass (OB; <i>M. cf. dolomieu</i> Ouachita River), and the Little River Bass (LRB; <i>M. cf. dolomieu </i> Little River).</font>

## Specific Aim: Data preparation, enumeration, and filtering
For this aim, we clean and filter the full genotype data for 487 black bass individuals, which was derived from the diagnostic SNP panel developed by Long et al. (2020) and prepare the data for analysis in Structure and NewHybrids (See Analysis 2).

Specifically, we filter the dataset based on three criteria:

  1) poor quality SNP loci (loci that failed to genotype in over 20% of samples);
  2) poor quality samples (samples that failed to genotype in over 20% of loci); and
  3) potential duplicate samples (samples that are greater than 95% identical across loci) 

## Phases of Analysis
### Phase 1: Data Preparation
### Phase 2: Data summarization

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(writexl)
```

## PHASE 1: DATA PREPARATION 

### STEP 1: Read in full dataset and generate metadata
In this step, we read in the raw data file (which includes sample metadata and SNP genotype data; `../raw_data/genotype_data.xlsx`), manipulate and clean the data, and generate separate data frames for the metadata (without SNP genotype data) and for the SNP genotype data (without metadata).

#### 1a: Read in, clean, and save full dataset. 
As part of data cleaning, we are omitting three samples representing Shoal Bass (<i>M. cataractae</i>) due to low sample size and because we are not explicitly interested in Shoal Bass hybridization or population structure.
 
<b>Before filtering:</b> <br>
<i>N</i><sub>(sample)</sub> = 487 <br>
<i>N</i><sub>(loci)</sub> = 192 <br>

##### 1a.1. Read in full dataset, convert characters to factors, and identify Shoal Bass samples to be removed from downstream analyses (first three rows of data); run the Rmd chunk below.

##### Read in and clean the dataset and identify Shoal Bass:
```{r}
# Read in raw genotype data
genotype_data <- read_excel("../raw_data/genotype_data.xlsx") 

# Convert characters to factors
genotype_data <- genotype_data %>%
  mutate(Sample_ID = factor(Sample_ID))

# Save raw genotype data
save(genotype_data, file = "data/processed_raw_data/gdata.Rda")

# Read in raw genotype data
metadata <- read_excel("../raw_data/metadata.xlsx") 

# Convert characters to factors
metadata <- metadata %>%
  mutate(Sample_ID = factor(Sample_ID)) %>%
  mutate(Sample_Num = factor(Sample_Num)) %>%
  mutate(Putative_Taxon = factor(Putative_Taxon)) %>%
  mutate(Vis_ID = factor(Vis_ID)) %>%
  mutate(River = factor(River)) %>%
  mutate(Location = factor(Location)) %>%
  mutate(Genetics_Num = factor(Genetics_Num)) %>%
  mutate(Structure_Num = factor(Structure_Num))

# Save raw metdata data
save(metadata, file = "data/processed_raw_data/mdata.Rda")

# Get list of shoal bass from metadata
shoal_bass <- metadata %>%
  filter(Putative_Taxon == "Shoal") %>%
  dplyr::select(Sample_ID)

# Convert column to rownames
shoal_bass <- column_to_rownames(shoal_bass, "Sample_ID")

# Save bad snps data for downstream filtering
save(shoal_bass, file = "data/filtering_data/shoal_bass.Rda")
```

##### 1a.2. Remove shoal bass samples from dataset; run the Rmd chunk below.

##### Remove shoal bass samples from full dataset:
```{r}
# Load shoal bass data
load("data/filtering_data/shoal_bass.Rda")

# Load raw genotype data
load("data/processed_raw_data/gdata.Rda")

# Load raw metadata
load("data/processed_raw_data/mdata.Rda")

# Remove shoal bass from genotype data
gdata_01_filter_shoal <- genotype_data[ ! genotype_data$Sample_ID %in% row.names(shoal_bass), ]

# Save genotype data (gdata) with Shoal Bass filter for downstream filtering steps
save(gdata_01_filter_shoal, file = "data/processed_raw_data/gdata_01_filter_shoal.Rda")

# Remove shoal bass from metadata
mdata_01_filter_shoal <- metadata[ ! metadata$Sample_ID %in% row.names(shoal_bass), ]

# Save metadata (mdata) with Shoal Bass filter for downstream filtering steps
save(mdata_01_filter_shoal, file = "data/processed_raw_data/mdata_01_filter_shoal.Rda")
```

<b>Filtering results:</b> <br>
<i>N</i><sub>(sample)</sub> = 484 <br>
<i>N</i><sub>(loci)</sub> = 192 <br>

In this step, we are counting the total number of samples (before and after filtering; see Step 1a above) across sample sites and obtaining counts of samples per species.

#### 1b: Count samples in the full, unfiltered dataset; run the Rmd chunk below:

##### Sample count summaries in the full dataset:
```{r}
## Load in full metadata (excluding shoal bass)
load("../filtering_analysis/data/processed_raw_data/mdata_01_filter_shoal.Rda")

## Get counts of individuals per species
mdata_01_filter_shoal %>%
  group_by(Putative_Taxon) %>%
  count()

## Get counts of individuals per river per species
mdata_01_filter_shoal %>%
  group_by(Putative_Taxon, River) %>%
  count()
```

<b>Summary of full dataset:</b> <br>

<b>Number of samples by taxon within genetic group</b>

<b>SPB:</b> 8 <br>

<b>SMB-C:</b> 476 <br>
<b>SMB:</b> 24<br>

<b>CIH:</b> 452 <br>
<b>NB:</b> 50 <br>
<b>LRB:</b> 243<br>
<b>OB:</b> 159<br>

<b>Number of samples by river</b>

<b>Big_Eagle_Creek:</b>	3		
<b>Blackfork_Creek:</b>	34		
<b>Cossatot_River_AR:</b>	33		
<b>Glover_River:</b>	55		
<b>Honia_Creek:</b>	23		
<b>Little_River:</b>	5		
<b>Mountain_Fork:</b>	15		
<b>Pero_Creek_Rolling_Fork_trib:</b>	1		
<b>Upper_Mountain_Fork:</b>	37		
<b>Western_Saline_River_AR:</b>	37	
<b>Baron_Fork:</b>	10		
<b>Caney_Creek:</b>	10		
<b>Honey_Creek:</b>	10		
<b>Lee_Creek:</b>	10		
<b>Spavinaw_Creek:</b>	10		
<b>Caddo_River_AR:</b>	40		
<b>Eastern_Saline_River_AR:</b>	40		
<b>Little_Missouri_River_AR:</b>	39		
<b>Ouachita_River_AR:</b>	40		
<b>Lake_Erie_Central_Basin_Ohio:</b>	4	
<b>Lake_Superior_Chequamegon_Bay:</b>	4		
<b>Oneida_Lake_NY:</b>	4		
<b>Skiatook_Lake:</b>	8		
<b>Smith_Fork_Creek:</b>	4		
<b>Glover_River:</b>	2		
<b>Illinois_River:</b>	6	

This information is the basis for Table 1 in the final manuscript.

### STEP 2: Assess data quality
In this step, we assess the quality of the raw genotype data by calculating the rate of missing genotype calls per locus across samples (i.e., for each single locus, how many genotypes are missing across the total number of individuals) and per sample across loci (i.e., for each single sample, how many genotypes are missing across the total number SNPs). We use these summary statistics to establish a baseline for determining an appropriate missing genotype call rate threshold for samples and loci to be applied in downstream filtering steps.

#### 2a: Assess missing genotype rates for loci across samples (per-locus genotype missingness).

##### 2a.1. Calculate per-locus genotype missingness; run the Rmd chunk below.

##### Calculate per-locus genotype missingness:
```{r}
# Load genotype data with Shoal Bass filter (gdata_01_filter_shoal.Rda)
load("data/processed_raw_data/gdata_01_filter_shoal.Rda")

# Get counts of "no-genotype" calls per site across samples
raw_snp_missing <- apply(gdata_01_filter_shoal[2:193], 2, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
raw_snp_missing <- as.data.frame(raw_snp_missing)

# Convert rownames (which are SNP Ids) to column and call it "snp_id"
raw_snp_missing <- rownames_to_column(raw_snp_missing, "snp_id")

# Calculate rate of missing genotypes per locus (number of missing genotypes divided by total number of samples (484 after shoal bass filtering))
raw_snp_missing <- raw_snp_missing %>%
  mutate(missing_rate = raw_snp_missing/484)

# Save SNP missingness data for summary analysis 
save(raw_snp_missing, file = "data/summary_data/raw_snp_missing.Rda")
```

#### 2a.2. Calculate summary statistics for per-locus missingness; run the Rmd chunk below:

##### Summarize per-locus missingness:
```{r}
# Load SNP missingness data 
load("data/summary_data/raw_snp_missing.Rda")

# Calculate average number of SNPs missing per locus across samples
raw_snp_missing %>%
  summarize(mean_snp_missing = mean(raw_snp_missing))

# Calculate average missingness rate per locus across samples
raw_snp_missing %>%
  summarize(mean_missing_rate = mean(missing_rate))

# Get range of SNP missingness
raw_snp_missing %>%
  arrange(desc(missing_rate))
```

<b>Summary of pre-filtering locus genotype missingness:</b> <br>

<b>Average number of genotypes missing:</b> 17.28125 <br>
<b>Average SNPs missingness rate:</b> 0.03570506<br>

#### 2b: Assess missing genotype rates for samples across loci (per-sample genotype missingness).

##### 2b.1. Calculate per-sample genotype missingness; run the Rmd chunk below.

##### Calculate per-sample genotype missingness:
```{r}
# Load genotype data with Shoal Bass filter (gdata_01_filter_shoal.Rda)
load("data/processed_raw_data/gdata_01_filter_shoal.Rda")

# Get counts of "no-genotype" calls per sample across loci
raw_sample_missing <- apply(gdata_01_filter_shoal, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe 
raw_sample_missing <- as.data.frame(raw_sample_missing)

# Bind with sample ID
raw_sample_missing <- cbind(gdata_01_filter_shoal$Sample_ID, raw_sample_missing)

# Change first column name
colnames(raw_sample_missing) <- c("sample_id", "raw_sample_missing")

# Calculate rate of missing genotypes per sample (number of missing genotypes divided by total number of SNP loci (186 after poor quality SNP filtering filtering))
raw_sample_missing <- raw_sample_missing %>%
  mutate(missing_rate = raw_sample_missing/192)

# Save sample missingness data for summary analysis
save(raw_sample_missing, file = "data/summary_data/raw_sample_missing.Rda")
```

#### 2b.2. Calculate summary statistics for per-sample genotype missingness; run the Rmd chunk below:

##### Summarize per-sample genotype missingness:
```{r}
# Load SNP missingness data 
load("data/summary_data/raw_sample_missing.Rda")

# Calculate average number of genotypes missing per sample across loci
raw_sample_missing %>%
  summarize(mean_sample_missing = mean(raw_sample_missing))

# Calculate average missingness rate per sample across loci
raw_sample_missing %>%
  summarize(mean_missing_rate = mean(missing_rate))

# Get range of SNP missingness
raw_sample_missing %>%
  arrange(desc(missing_rate))
```

<b>Summary of pre-filtering sample genotype missingness:</b> <br>

<b>Average number of genotypes missing:</b> 6.855372 <br>
<b>Average per-sample missingness rate:</b> 0.03570506<br>

### STEP 3. Filter loci with poor genotype quality (high missingness rates across samples ).
Here, we identify loci with missing genotype call rates of >20% across samples and samples with missing genotype call rates of >20% across loci. 

#### 3a. Identify loci with per-locus missing genotype call rate higher than 20%; run the Rmd chunk below:

##### Identify loci with poor per-locus genotype call rate:
```{r}
# Load locus missingness data 
load("data/summary_data/raw_snp_missing.Rda")

# Generate list of poor quality snp markers
bad_snps <- raw_snp_missing %>%
  filter(missing_rate > 0.20) %>%
  select(snp_id)

# Convert column to rownames
bad_snps <- column_to_rownames(bad_snps, "snp_id")

# Save bad snps data for downstream filtering
save(bad_snps, file = "data/filtering_data/bad_snps.Rda")
```

#### 3b. Remove loci with high missing genotype call rates across samples; run the Rmd chunk below.

##### Remove poor quality loci from the raw data:
```{r}
# Load bad locus data
load("data/filtering_data/bad_snps.Rda")

# Load data filtered for shoal bass (see Step 1a.1 above)
load("data/processed_raw_data/gdata_01_filter_shoal.Rda")

# Remove 15 poor quality SNPs from genotype data
gdata_02_filter_snps <- gdata_01_filter_shoal[ , -which(names(gdata_01_filter_shoal) %in% row.names(bad_snps)), drop=F]

# Save genotype data (gdata) with SNP filter for downstream filtering steps
save(gdata_02_filter_snps, file = "data/processed_raw_data/gdata_02_filter_snps.Rda")
```

<b>Filtering results:</b> <br>

<b>The following LOCI were removed from further analysis:</b> <br>
NEO_locus17285_283<br>
NEO_locus17912_112<br>				
OULR_locus10488_266<br>				
OUOU_locus11827_147<br>				
OUOU_locus6374_33<br>		
OULR_locus1841_280<br>

<i>N</i><sub>(sample)</sub> = 484 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

#### 3b: Recalculate per-sample genotype missingness after filtering poor quality loci, identify poor quality samples, and remove them from the filtered dataset; run the Rmd chunk below.

##### Recalculate per-sample genotype missingness in the filtered dataset (after removal of poor quality loci):
```{r}
# Load genotype data filtered for low quality SNPs
load("data/processed_raw_data/gdata_02_filter_snps.Rda")

# Load metadata filtered for shoal bass
load("data/processed_raw_data/mdata_01_filter_shoal.Rda")

# Get counts of "no-genotype" calls per sample across loci
sample_missing <- apply(gdata_02_filter_snps, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe 
sample_missing <- as.data.frame(sample_missing)

# Bind per-sample missingness data to metadata
sample_missing <- cbind(mdata_01_filter_shoal, sample_missing)

# Calculate rate of missing genotypes per sample (number of missing genotypes divided by total number of SNP loci (186 after poor quality SNP filtering filtering))
sample_missing <- sample_missing %>%
  mutate(missing_rate = sample_missing/186)

# View genotype missingness in descending order for filtering 
sample_missing %>%
  arrange(desc(missing_rate))

# Generate list of poor quality samples
bad_samples <- sample_missing %>%
  filter(missing_rate > 0.20) %>%
  select(Sample_ID)

# Convert column to rownames 
bad_samples <- column_to_rownames(bad_samples, "Sample_ID")

# Remove 6 poor quality samples from genotype data
gdata_03_filter_samples <- gdata_02_filter_snps[ ! gdata_02_filter_snps$Sample_ID %in% row.names(bad_samples), ]

# Save genotype data (gdata) with SNP filter for downstream filtering steps
save(gdata_03_filter_samples, file = "data/processed_raw_data/gdata_03_filter_samples.Rda")

# Remove poor quality samples from metadata
mdata_03_filter_samples <- mdata_01_filter_shoal[ ! mdata_01_filter_shoal$Sample_ID %in% row.names(bad_samples), ]

# Save metadata (mdata) with sample filter for downstream filtering steps
save(mdata_03_filter_samples, file = "data/processed_raw_data/mdata_03_filter_samples.Rda")
```

<b>The following SAMPLES were removed from further analysis:</b> <br>
OUACH007 <br>
OUACH008 <br>
G1018-03-OSU_SKIA-4 <br>
G1018-03-OSU_GRSPB-001 <br>
G1018-03-OSU_SPVW-004 <br>
OUACH015 <br>

<b>After filtering, the data set consists of:</b> <br>
<i>N</i><sub>(sample)</sub> = 478 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

### STEP 4: Filter the dataset for potential (likely) duplicate samples based on percent identity.
The Center of Aquaculture Technologies (CAT), who genotyped our black bass samples, used percent identity to detect potential duplicate samples in the dataset. They identified 14 samples as potential duplicates with a percent identity over 95% (0.95). 

<b>List of duplicates (% identity):</b> <br>
GLVR-011 & GLVR-022 (98.8) <br>
UMF_045 & UMF035 (100) <br>
G1018-03-OSU_CHA-038 & G1018-03-OSU_CHA-040 (Shoal Bass) (100) <br>
G1018-03-OSU_GLVR-006	& GLVR-008 (95.6) <br>
UMF006 & UMF012	(100) <br>
G1018-03-OSU_BFC-023 & BFC061 (96.6) <br>
G1018-03-OSU_GLVR-005	& G1018-03-OSU_GLVR-024	(96.6) <br>

Two of the duplicate samples (G1018-03-OSU_CHA-038 & G1018-03-OSU_CHA-040) were Shoal Bass and were already removed in step 1a.1 above. To omit inference on duplicate samples in downstream analyses, we removed one of the duplicates from each of the six remaining pairs of duplicates. Specifically, we removed the following samples:

<b>Duplicate samples removed:</b> <br>
GLVR-011 <br>
UMF_045 <br>
G1018-03-OSU_GLVR-006 <br>
UMF006 <br>
G1018-03-OSU_BFC-023 <br>
G1018-03-OSU_GLVR-005 <br>

##### 4a: Remove duplicate samples from genotype data and metadata; run the Rmd chunk below:

##### Omit duplicate samples:
```{r}
# Generate list of poor quality samples
duplicate_samples <- c("GLVR-011","UMF_045", "G1018-03-OSU_GLVR-006", "UMF006", "G1018-03-OSU_BFC-023", "G1018-03-OSU_GLVR-005")

# Remove 6 poor quality samples from genotype data
gdata_04_filter_duplicates <- gdata_03_filter_samples[ ! gdata_03_filter_samples$Sample_ID %in% duplicate_samples, ]

# Save genotype data (gdata) with duplicate filter for downstream filtering steps
save(gdata_04_filter_duplicates, file = "data/processed_raw_data/gdata_04_filter_duplicates.Rda")

# Remove duplicates from metadata
mdata_04_filter_duplicates <- mdata_03_filter_samples[ ! mdata_03_filter_samples$Sample_ID %in% duplicate_samples, ]

# Save metadata (mdata) with duplicate filter for downstream filtering steps
save(mdata_04_filter_duplicates, file = "data/processed_raw_data/mdata_04_filter_duplicates.Rda")

# Merge gdata and mdata to create the fully filtered dataset
filtered_data <- merge(mdata_04_filter_duplicates,
                       gdata_04_filter_duplicates, 
                       by = "Sample_ID")

# Save the fully filtered data for downstream analyses
save(filtered_data, file = "data/processed_raw_data/filtered_data.Rda")
```

<b>After filtering, the data set consists of:</b> <br>
<i>N</i><sub>(sample)</sub> = 472 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

### STEP 5: Summarize genotype missingness in fully filtered data.
In this step, we summarize genotype missingness for loci across samples and for samples across loci to determine the effectiveness of the 20% missing call rate threshold. Specifically, we assess missingness of loci and samples hierarchically according to our genetic analysis framework, i.e., 1) we isolate only Spotted Bass loci and assess missingness for all individuals; 2) we isolate only Smallmouth Bass loci and assess missingness for all individuals except for Spotted Bass; and 3) we isolate only CIH species (Neosho, Little River, and Ouachita Basses) and assess missingness for all CIH individuals. We also assess whether genotype missingness is clustered by these genetic groups.

#### 5a. Calculate and summarize per-locus and per-sample genotype missingness for the fully filtered dataset.
In this step, we recalculate and summarize per-locus and per-sample genotype missingness for the fully filtered dataset, and these statistics are reported in the maintext Results section of the final manuscript.

##### 5a.1. Calculate per-locus genotype missignness on filtered data; run the Rmd chunk below.

##### Calculate and summarize per-locus genotype missingness on filtered data:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get counts of "no-genotype" calls per site across samples
filtered_snp_missing <- apply(filtered_data[14:199], 2, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
filtered_snp_missing <- as.data.frame(filtered_snp_missing)

# Convert rownames (which are SNP Ids) to column and call it "snp_id"
filtered_snp_missing <- rownames_to_column(filtered_snp_missing, "snp_id")

# Rename columns
colnames(filtered_snp_missing) <- c("snp_id", "snp_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
filtered_snp_missing <- filtered_snp_missing %>%
  mutate(missing_rate = snp_missing/472)

# Calculate average number of missing genotypes across individuals for each locus
filtered_snp_missing %>%
  summarize(mean_snps_missing = mean(snp_missing))

# Calculate average genotype missingness rate across individuals for each locus
filtered_snp_missing %>%
  summarize(mean_missing_rate = mean(missing_rate))

# Get the range of per-locus genotype missigness
filtered_snp_missing %>%
  arrange(desc(missing_rate))
```

<b>Summary of post-filtering per-locus genotype missingness:</b> <br>

<b>Average number of genotypes missing:</b> 4.887097 <br>
<b>Average per-locus missingness rate:</b> 0.01035402<br>

#### 5a.2. Calculate per-sample missingness on filtered data; run the Rmd chunk below.

##### Calculate and summarize per-sample genotype missingness on filtered data:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get counts of "no-genotype" calls per site across samples
filtered_sample_missing <- apply(filtered_data, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
filtered_sample_missing <- as.data.frame(filtered_sample_missing)

# Rename columns
colnames(filtered_sample_missing) <- c("sample_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
filtered_sample_missing <- filtered_sample_missing %>%
  mutate(missing_rate = sample_missing/186)

# Calculate average number of missing genotypes across loci for each individual
filtered_sample_missing %>%
  summarize(mean_samples_missing = mean(sample_missing))

# Calculate average genotype missingness rate across loci for each individual
filtered_sample_missing %>%
  summarize(mean_missing_rate = mean(missing_rate))

# Get the range of per-sample genotype missigness
filtered_sample_missing %>%
  arrange(desc(missing_rate))
```

<b>Summary of post-filtering sample genotype missingness:</b> <br>

<b>Average number of genotypes missing:</b> 1.925847 <br>
<b>Average SNP missingness rate:</b> 0.01035402<br>

#### 5b. Calculate and compare per-sample genotype missingness for SPB and SMBC. 
In this step, we calculate missingness across individuals for only the SPB and SMBC dataset (all individuals but only SPB loci) to assess the potential for biased inference due to missing data.

##### 5b.1. Generate per-sample missigness data for SPB vs. SMB-C; run the Rmd chunk below.

##### Separate individual analysis datasets:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get filtered genotype data only along with species ID
filtered_genotypes <- filtered_data[,c(5,14:199)]

# Get SPB and SMB-C data
spb_smbc <- filtered_genotypes

# Get species labels 
spb_smbc_spp <- data.frame(spb_smbc[,c(1)])

# Change column name 
colnames(spb_smbc_spp) <- c("taxon")

# Get SPB loci only
spb_smbc <- spb_smbc %>%
  dplyr::select(matches("SM_SP"))

# Rejoin species labels with genotype data
spb_smbc <- cbind(spb_smbc_spp, spb_smbc)

# Get counts of "no-genotype" calls per site across samples
spb_smbc_sample_missing <- apply(spb_smbc, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
spb_smbc_sample_missing <- as.data.frame(cbind(spb_smbc_spp, spb_smbc_sample_missing))

# Rename columns
colnames(spb_smbc_sample_missing) <- c("taxon", "sample_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
spb_smbc_sample_missing <- spb_smbc_sample_missing %>%
  mutate(missing_rate = sample_missing/15)

# Create group metadata label
spb_smbc_sample_missing$group <- with(spb_smbc_sample_missing, ifelse(taxon=="SPB","SPB",
                                                                      ifelse(taxon != "SPB","SMB-C","done")))

# Save SPB and SMB-C data for downstream analysis
save(spb_smbc_sample_missing, file = "data/summary_data/spb_smbc_sample_missing.Rda")

spb_smbc %>%
  filter(taxon == "SPB")
```

##### 5b.2. Summarize per-sample missigness data for SPB vs. SMB-C; run the Rmd chunk below.

##### Summarize per-sample missignness data for SPB and SMB-C:
```{r}
# Load SMB and SMB-C missingness data
load("data/summary_data/spb_smbc_sample_missing.Rda")

ave_missing <- spb_smbc_sample_missing %>%
  group_by(group) %>%
  summarize(mean_sample_missing = mean(sample_missing),
            mean_missing_rate = mean(missing_rate))
  

nonzero <- spb_smbc_sample_missing %>%
 filter(sample_missing > 0)

set.seed(432431)
nreps <- 1e3
t_stats <- numeric(nreps)
obs_glm <- glm(sample_missing ~ group, family = "gamma", data=spb_smbc_sample_missing)
obs_t <- summary(obs_glm)$coefficients[2,3]
t_stats[1] <- obs_t

for (ii in 2:nreps) {
  rand_glm <- glm(sample(sample_missing) ~ group, family = "poisson", data=spb_smbc_sample_missing)
  rand_t <- summary(rand_glm)$coefficients[2,3]
  t_stats[ii] <- rand_t
}

t_stats <- as.data.frame(t_stats)

thist <- ggplot(t_stats, aes(t_stats)) + geom_histogram(bins = 100) 

mean(t_stats >= t_stats[1,1])
```

#### 5c. Calculate and compare per-sample genotype missingness for SMB and CIH 
In this step, we calculate missingness across individuals for only the SMB and CIH dataset (no SPB and only SMB loci) to assess the potential for biased inference due to missing data.

##### 5c.1. Generate per-sample missigness data for SMB vs. CIH; run the Rmd chunk below.

##### Separate individual analysis datasets:
```{r}
# Get SMB and CIH data
smb_cih <- filtered_genotypes %>%
  filter(Putative_Taxon != "SPB")

# Get species labels 
smb_cih_spp <- data.frame(smb_cih[,c(1)])

# Change column name 
colnames(smb_cih_spp) <- c("taxon")

# Get SPB loci only
smb_cih <- smb_cih %>%
  dplyr::select(matches("N_SM"))

# Rejoin species labels with genotype data
smb_cih <- cbind(smb_cih_spp, smb_cih)

# Get counts of "no-genotype" calls per site across samples
smb_cih_sample_missing <- apply(smb_cih, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
smb_cih_sample_missing <- as.data.frame(cbind(smb_cih_spp, smb_cih_sample_missing))

# Rename columns
colnames(smb_cih_sample_missing) <- c("taxon", "sample_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
smb_cih_sample_missing <- smb_cih_sample_missing %>%
  mutate(missing_rate = sample_missing/33)

# Create group metadata label
smb_cih_sample_missing$group <- with(smb_cih_sample_missing, ifelse(taxon=="SMB","SMB",
                                                                    ifelse(taxon != "SMB","CIH","done")))

# Save SPB and SMB-C data for downstream analysis
save(smb_cih_sample_missing, file = "data/summary_data/smb_cih_sample_missing.Rda")
```

##### 5c.2. Summarize per-sample missigness data for SMB vs. CIH; run the Rmd chunk below.

##### Summarize per-sample missignness data for SMB vs. CIH:
```{r}
# Load SMB and SMB-C missingness data
load("data/summary_data/smb_cih_sample_missing.Rda")

ave_missing <- smb_cih_sample_missing %>%
  group_by(group) %>%
  summarize(mean_sample_missing = mean(sample_missing),
            mean_missing_rate = mean(missing_rate))
  

set.seed(432431)
nreps <- 1e4
t_stats <- numeric(nreps)
obs_glm <- glm(sample_missing ~ group, family = "poisson", data=smb_cih_sample_missing)
obs_t <- summary(obs_glm)$coefficients[2,3]
t_stats[1] <- obs_t

for (ii in 2:nreps) {
  rand_glm <- glm(sample(sample_missing) ~ group, family = "poisson", data=smb_cih_sample_missing)
  rand_t <- summary(rand_glm)$coefficients[2,3]
  t_stats[ii] <- rand_t
}

t_stats <- as.data.frame(t_stats)

thist <- ggplot(t_stats, aes(t_stats)) + geom_histogram(bins = 100) 

mean(t_stats >= t_stats[1,1])
```

```{r}

# Get CIH data
cih <- filtered_genotypes %>%
  filter(Putative_Taxon != "SPB" & Putative_Taxon != "SMB")

# Get species labels 
cih_spp <- data.frame(cih[,c(1)])

# Change column name 
colnames(cih_spp) <- c("taxon")

# Get SPB loci only
cih <- cih %>%
  dplyr::select(matches("NEO"),
         matches("OUOU"),
         matches("OULR"))

# Rejoin species labels with genotype data
cih <- cbind(cih_spp, cih)




# Create group metadata label
filtered_sample_missing$spb_smbc <- with(filtered_sample_missing, ifelse(Putative_Taxon=="SPB","SPB",
                                                                         ifelse(Putative_Taxon != "SPB","SMB-C","done")))

filtered_sample_missing$smb_cih <- with(filtered_sample_missing, ifelse(Putative_Taxon=="SMB","SMB",
                                                                         ifelse(Putative_Taxon == "SPB","NA",
                                                                                ifelse(Putative_Taxon =="LRB" | Putative_Taxon == "OB" | Putative_Taxon == "NB", "CIH", "done"))))


# Create data frame with average and standard deviation of missingness rate
ave_sample_missing <- filtered_sample_missing %>%
  group_by(Putative_Taxon) %>%
  summarize(mean = mean(sample_missing), 
            sd = sd(sample_missing), 
            var = sd^2, 
            n = n(), 
            se = sd/sqrt(n)) %>%
  as.data.frame()

ggplot(ave_sample_missing, aes(x = Putative_Taxon, y = mean, fill = "grey")) + 
  geom_bar(color = "black", stat = "identity", show.legend = F) +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2,
                 position=position_dodge(.9)) +
  theme_set(theme_cowplot(12)) +
  labs(x = "Hierarchical group", y = "Missing genotypes across loci (%)") +
  scale_fill_manual(values = c("grey")) +
  theme(axis.title = element_text(size = 20)) + 
  theme(axis.text = element_text(size = 15)) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_blank()) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

```

##### 5b.1. Calculate per-sample genotype missigness for SPB and SMBC; run the Rmd chunk below.

##### Calculate and summarize genotype missingness for loci across samples:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get only SPB SNPS
spb_data <- filtered_data %>%
  dplyr::select(matches("SM_SP"))

# Get counts of "no-genotype" calls per site across samples
spb_smbc_sample_missing <- apply(spb_data, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
spb_smbc_sample_missing <- as.data.frame(spb_smbc_sample_missing)

# Modify column names
colnames(spb_smbc_sample_missing) <- c("sample_missing")

# Bind with sample ID
spb_smbc_sample_missing <- cbind(filtered_data[,c(1,5)], spb_smbc_sample_missing)

# Calculate rate of missing genotypes per locus for the filtered dataset
spb_smbc_sample_missing <- spb_smbc_sample_missing %>%
  mutate(missing_rate = sample_missing/15)

# Save final sample missingness data 
save(spb_smbc_sample_missing, file = "data/filtering_data/spb_smbc_sample_missing.Rda")
```

##### 5b.2. Compare genotype missingness across loci between SPB and SMBC groups to assess potential bias.
In this step, we compare average missingness across individuals (calculated in Step 5b.1 above) among genetic groups to determine whether missing genotypes may influence inference of hybridization.

##### Summarize genotype missingness by genetic group and plot :
```{r}
# Load sample missing data 
load("data/filtering_data/spb_smbc_sample_missing.Rda")

# Create group metadata label
spb_smbc_sample_missing$group <- with(spb_smbc_sample_missing, ifelse(Putative_Taxon=="SPB","SPB",
                                                                      ifelse(Putative_Taxon != "SPB","SMB-C","done")))

# Create data frame with average and standard deviation of missingness rate
spb_smbc_ave_missing <- spb_smbc_sample_missing %>%
  group_by(group) %>%
  summarize(mean = mean(sample_missing), 
            sd = sd(sample_missing), 
            var = sd^2, 
            n = n(), 
            se = sd/sqrt(n)) %>%
  as.data.frame()

# Assess difference in missingness between genetic groups using linear model
mod <- glm(sample_missing ~ group,
           data = spb_smbc_sample_missing)

# Summarize model coefficients
summary(mod)

# Plot missingness by group
pdf("figures/spb_smbc_missing.pdf", width=6, height=6)

ggplot(spb_smbc_ave_missing, aes(x = group, y = mean, fill = "grey")) + 
  geom_bar(color = "black", stat = "identity", show.legend = F) +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2,
                 position=position_dodge(.9)) +
  theme_set(theme_cowplot(12)) +
  labs(x = "Hierarchical group", y = "Missing genotypes across loci (%)") +
  scale_fill_manual(values = c("grey")) +
  theme(axis.title = element_text(size = 20)) + 
  theme(axis.text = element_text(size = 15)) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_blank()) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

#### 5c. Calculate and compare per-sample genotype missingness for SMB and CIH 
In this step, we calculate missingness across individuals for only the SMB and CIH dataset (all individuals but except SPB, and only SMB loci) to assess the potential for biased inference due to missing data.

##### 5c.1. Calculate per-sample genotype missigness for SMB and CIH; run the Rmd chunk below.

##### Calculate and summarize genotype missingness for loci across samples:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get only SPB SNPS
smb_data <- filtered_data %>%
  filter(Putative_Taxon != "SPB") %>%
  dplyr::select(Putative_Taxon, matches("N_SM"))

# Get counts of "no-genotype" calls per site across samples
smb_cih_sample_missing <- apply(smb_data, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
smb_cih_sample_missing <- as.data.frame(smb_cih_sample_missing)

# Bind with sample ID
smb_cih_sample_missing <- cbind(smb_data$Putative_Taxon, smb_cih_sample_missing)

# Modify column names
colnames(smb_cih_sample_missing) <- c("Putative_Taxon","sample_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
smb_cih_sample_missing <- smb_cih_sample_missing %>%
  mutate(missing_rate = sample_missing/33)

# Save final sample missingness data 
save(smb_cih_sample_missing, file = "data/filtering_data/smb_cih_sample_missing.Rda")
```

##### 5c.2. Compare genotype missingness across loci between SMB and CIH groups to assess potential bias.
In this step, we compare average missingness across individuals (calculated in Step 5c.1 above) among genetic groups to determine whether missing genotypes may influence inference of hybridization.

##### Summarize genotype missingness by genetic group and plot :
```{r}
# Load sample missing data 
load("data/filtering_data/smb_cih_sample_missing.Rda")

# Create group metadata label
smb_cih_sample_missing$group <- with(smb_cih_sample_missing, ifelse(Putative_Taxon=="SMB","SMB",
                                                                    ifelse(Putative_Taxon != "SMB","CIH","done")))

# Create data frame with average and standard deviation of missingness rate
smb_cih_ave_missing <- smb_cih_sample_missing %>%
  group_by(group) %>%
  summarize(mean = mean(sample_missing), 
            sd = sd(sample_missing), 
            var = sd^2, 
            n = n(), 
            se = sd/sqrt(n)) %>%
  as.data.frame()

# Assess difference in missingness between genetic groups using linear model
mod <- glm(sample_missing ~ group,
           data = smb_cih_sample_missing)

# Summarize model coefficients
summary(mod)

ggplot(spb_smbc_sample_missing, aes(missing_rate)) + 
  geom_histogram(binwidth = 0.05)

# Plot missingness by group
pdf("figures/smb_cih_missing.pdf", width=6, height=6)

ggplot(smb_cih_ave_missing, aes(x = group, y = mean, fill = "grey")) + 
  geom_bar(color = "black", stat = "identity", show.legend = F) +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2,
                 position=position_dodge(.9)) +
  theme_set(theme_cowplot(12)) +
  labs(x = "Hierarchical group", y = "Missing genotypes across loci (%)") +
  scale_fill_manual(values = c("grey")) +
  theme(axis.title = element_text(size = 20)) + 
  theme(axis.text = element_text(size = 15)) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_blank()) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

#### 5d. Calculate and compare per-sample genotype missingness for CIH species
In this step, we calculate missingness across individuals for only the SMB and CIH dataset (all individuals except SMB and SPB, and only CIH loci) to assess the potential for biased inference due to missing data.

##### 5d.1. Calculate per-sample genotype missigness for CIH species; run the Rmd chunk below.

##### Calculate and summarize genotype missingness for loci across samples:
```{r}
# Load fully filtered data
load("data/processed_raw_data/filtered_data.Rda")

# Get only SPB SNPS
cih_data <- filtered_data %>%
  filter(Putative_Taxon != "SPB" & Putative_Taxon != "SMB") %>%
  dplyr::select(Putative_Taxon, matches("NEO"),
         matches("OUOU"),
         matches("OULR"))

# Get counts of "no-genotype" calls per site across samples
cih_sample_missing <- apply(cih_data, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
cih_sample_missing <- as.data.frame(cih_sample_missing)

# Bind with sample ID
cih_sample_missing <- cbind(cih_data$Putative_Taxon, cih_sample_missing)

# Modify column names
colnames(cih_sample_missing) <- c("Putative_Taxon","sample_missing")

# Calculate rate of missing genotypes per locus for the filtered dataset
cih_sample_missing <- cih_sample_missing %>%
  mutate(missing_rate = sample_missing/138)

# Save final sample missingness data 
save(cih_sample_missing, file = "data/filtering_data/cih_sample_missing.Rda")
```

##### 5c.2. Compare genotype missingness across loci between SMB and CIH groups to assess potential bias.
In this step, we compare average missingness across individuals (calculated in Step 5c.1 above) among genetic groups to determine whether missing genotypes may influence inference of hybridization.

##### Summarize genotype missingness by genetic group and plot :
```{r}
# Load sample missing data 
load("data/filtering_data/cih_sample_missing.Rda")

# Create data frame with average and standard deviation of missingness rate
cih_ave_missing <- cih_sample_missing %>%
  group_by(Putative_Taxon) %>%
  summarize(mean = mean(sample_missing), 
            sd = sd(sample_missing), 
            var = sd^2, 
            n = n(), 
            se = sd/sqrt(n)) %>%
  as.data.frame()

# Assess difference in missingness between genetic groups using linear model
mod <- glm(sample_missing ~ Putative_Taxon,
           data = cih_sample_missing)

# Summarize model coefficients
summary(mod)

# Plot missingness by group
pdf("figures/cih_missing.pdf", width=6, height=6)

ggplot(cih_ave_missing, aes(x = Putative_Taxon, y = mean, fill = "grey")) + 
  geom_bar(color = "black", stat = "identity", show.legend = F) +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2,
                 position=position_dodge(.9)) +
  theme_set(theme_cowplot(12)) +
  labs(x = "Hierarchical group", y = "Missing genotypes across loci (%)") +
  scale_fill_manual(values = c("grey")) +
  theme(axis.title = element_text(size = 20)) + 
  theme(axis.text = element_text(size = 15)) + 
  theme(axis.line = element_line(color = "black")) +
  theme(axis.ticks = element_blank()) +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))

dev.off()
```

### STEP 5: Convert SNP genotype data to genepop compatible format.
Genepop data consists of genotypes in six digit format, where each allele is a three digit number. Here, we convert SNP nucleotides into two three-digit alleles arbitrarily as follows):

<b>SNP conversion codes:</b> <br>
<b>C</b>: 100 <br>
<b>G</b>: 110 <br>
<b>T</b>: 120 <br>
<b>A</b>: 130 <br>
<b>Missing genotypes</b>: 000 <br>
<b>":"</b> is dropped <br>

e.g., a heterozygous genotype of "A:C" converts to "130100"; a homozygous genotpye of "T:T" converts to "120120"

#### 5a: Load genotype data with all data filters ("04_filter_duplicates") and convert SNP nucleotide data to genepop compatible digits; run the Rmd chunk below:

##### Convert and generate SNP genotype data:
```{r}
# Load "clean_data" from .Rda file
load("data/processed_raw_data/gdata_04_filter_duplicates.Rda")

# Get genotypes only from all data
gdata_05_converted <- gdata_04_filter_duplicates[,2:187]

## Convert SNPs to genepop compatible format
gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub("C", "100", x) })) 

gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub("G", "110", x) }))

gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub("T", "120", x) }))

gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub("A", "130", x) }))

gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub("unknown", "000000", x) })) ## Missing data are demarcated as '000000'

gdata_05_converted <- data.frame(lapply(gdata_05_converted, function(x) {
  gsub(":", "", x) }))

gdata_05_converted <- cbind(gdata_04_filter_duplicates[,1], gdata_05_converted)

# Save SNP genotype data as .Rda file to be loaded in downstream analysis
save(gdata_05_converted, file = "data/processed_raw_data/gdata_05_converted.rda")
```

##### 5a.1. The SNP genotype data is saved as the .Rda file `data/processed_raw_data/gdata_05_converted.rda`, which can be loaded independently for downstream analyses. The associated R object is called "gdata_05_converted".

### STEP 6: Join filtered and converted genotype data and metadata for downstream Structure and NewHybrids analyses.
In this step, we join the fully filtered genotype data and metadata for all downstream analyses.

#### 6a: Load and join genotype data and metadata; run the Rmd chunk below.

##### Join metadata and genotype data into full dataset:
```{r}
# Load "genotype_data" from .Rda file
load("data/processed_raw_data/gdata_05_converted.Rda")

# Load "metadata" from .Rda file
load("data/processed_raw_data/mdata_04_filter_duplicates.Rda")

# cbind metadata and genotype data, omitting the redundant "Sample_ID" column
full_data <- cbind(mdata_04_filter_duplicates, gdata_05_converted[,2:187])

# save the full dataset for future analyses
save(full_data, file = "data/processed_raw_data/full_data.rda")
```

##### 6a.1. The cleaned full dataset is saved as the .Rda file `data/processed_raw_data/full_data.rda`, which can be loaded independently for downstream analyses. The associated R object is called "full_data".

## ------------------------ END OF PHASE 1: DATA PREPARATION ----------------------- ##


## PHASE 2: DATA SUMMARIZATION
In this phase of the analysis, we are summarizing the full dataset to determine the total number of samples, the number of sites covered, the number of samples per species, etc., for inclusion in the final manuscript.

### STEP 1: Sample summary counts
In this step, we are counting the total number of samples after all filtering steps across sample sites and obtaining counts of samples per species.

#### 1a: Count samples in the full, filtered dataset; run the Rmd chunk below:

##### Sample count summaries in the filtered dataset:
```{r}
# Load full data
load("data/processed_raw_data/full_data.rda")

## Get counts of individuals per species
full_data %>%
  group_by(Putative_Taxon) %>%
  count()

## Get counts of individuals per river per species
full_data %>%
  group_by(Putative_Taxon, River) %>%
  count()
```
These data are the basis for Table 1 in the manuscript.

## ------------------------ END OF PHASE 2: DATA SUMMARIZATION -------------------------- ##

## ------------------------ END OF ANALYSIS 1: FILTERING ANALYSIS ----------------------- ##
