---
<<<<<<< HEAD
title: "SNOLH Structure Analysis"
=======
title: "Structure and Hybrid Assignment Analysis"
>>>>>>> 01b1ed48716f80e224cfd81f79a4944bd3f2a21c
author: "Joe Gunn"
date: "2022-07-28"
output: html_document
---

<<<<<<< HEAD
# Project: Smallmouth, Neosho, Ouachita, and Little River Hybridization (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Smallmouth Bass (<i>Micropterus dolomieu</i>), the newly elevated Neosho Bass (<i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin (the Ouachita and Little River Basses)</font>

## Specific Aim: Hierarchical structure analysis with diagnostic markers
<font size="+1">For this aim, we are using Bayesian clustering analysis with STRUCTURE to assess patterns and levels of hybridization among and within black bass species in the Interior Highlands, with emphasis on populations in the Arkansas River Basin and Ouachita River Basin. We obtained SNP marker data from from Spotted Bass (<i>M. punctulatus</i>), Smallmouth Bass (<i>M. dolomieu</i>), Neosho Bass (<i>M. velox</i>), and two potentially distinct species in the Ouachita River Basin (Ouachita Bass and Little River Bass) which were known to be diagnostic of these species (Long et al. 2021).

### Phase 1: Data preparation
### Phase 2: Population structure analysis
### Phase 3: 
=======
# Project: Interior Highlands Hybridization: Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Spotted Bass (<i>Micropterus punctulatus</i>), Smallmouth Bass (<i>M. dolomieu</i>), the newly elevated Neosho Bass (<i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin (the Ouachita and Little River Basses)</font>

## Specific Aim: Hierarchical structure analysis with diagnostic markers
For this aim, we are using Bayesian clustering analysis with STRUCTURE to assess patterns and levels of hybridization among and within black bass species in the Interior Highlands, with emphasis on populations in the Arkansas River Basin and Ouachita River Basin. We obtained SNP marker data from from Spotted Bass (<i>M. punctulatus</i>), Smallmouth Bass (<i>M. dolomieu</i>), Neosho Bass (<i>M. velox</i>), and two potentially distinct species in the Ouachita River Basin (Ouachita Bass and Little River Bass) which were known to be diagnostic of these species (Long et al. 2021).

## Phases of Analysis
### Phase 1: Data preparation
### Phase 2: Structure and hybrid assignment
>>>>>>> 01b1ed48716f80e224cfd81f79a4944bd3f2a21c

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(writexl)
library(genepopedit)
library(pophelper)
```

## PHASE 1: DATA PREPARATION 

### STEP 1: Read in full dataset and generate metadata
In this step, we are reading in the raw data file (which includes sample metadata and SNP genotype data; 'snolh_all_data.xlsx'), manipulating and cleaning the data, and generating separate data frames for the metadata (without SNP genotype data) and for the SNP genotype data (without metadata). We will then be able to break the metadata and SNP data down into separate subsets of species/populations/markers, etc., and these subsets can be pieced together for subsequent hierarchical analyses.

### ***Skip to Step 1d to load in raw data objects immediately***
Steps 1a - 1c should be initially run to generate saved .Rda files. Once these files have been saved, there is no need to run Steps 1a - 1c again. We can skip directly to Step 1d to load in the data.

### ***Skip to Step 3b to load in processed genotype data objects (for structure analysis) immediately***
Steps 2a - 3a should be initially run after loading in raw data objects (see above) to generate saved .Rda files. Once these files have been saved, there is no need to run Steps 2a - 3a again. We can skip directly to Step 3b to load in the data.

#### 1a: Read in full dataset; Run the Rmd chunk below to read in, clean, and save full dataset. 

##### IMPORTANT NOTE ON DATA CLEANING: As part of data cleaning, we are omitting three samples representing Shoal Bass (<i>M. cataractae</i>) due to low sample size and because we are not explicitly interested in Shoal Bass hybridization or population structure. Additionally, we are removing 6 SNP markers which were shown to have low call rates across samples. Additionally, we are removing 2 samples with low genotyping rates across SNPs.

##### The following SNP LOCI were removed from further analysis:

<b>Markers with <50% call rates across samples:</b> <br>
OULR_locus1841_280 <br>
NEO_locus17912_112 <br>
OULR_locus10488_266 <br>

<b>Markers with ~60% call rates across samples:</b> <br>
OUOU_locus11827_147
NEO_locus17285_283
OUOU_locus6374_33

##### The following SAMPLES were removed from further analysis:

<b>Failed samples:</b> <br>
OUACH007
OUACH008

##### Read in, clean, and save full dataset: `data/raw_data/snolh_all_data.xlsx`
```{r}
# Read in raw dataset, with metadata appended
raw_data <- read_excel("data/raw_data/snolh_all_data.xlsx") 

# Convert characters to factors
clean_data <- raw_data %>%
  mutate(Ref_samps = factor(Ref_samps), 
         Sample_Num = factor(Sample_Num), 
         Putative_Taxon = factor(Putative_Taxon),
         Vis_ID = factor(Vis_ID),
         River = factor(River),
         Location = factor(Location),
         Date = factor(Date),
         Genetics_Num = factor(Genetics_Num),
         Structure_Num = factor(Structure_Num),
         Sample_ID = factor(Sample_ID))

# Omit the 3 shoal bass samples in the first 3 rows.
clean_data <- clean_data[-c(1:3),] 

# Omit 6 poor performing loci 
clean_data <- clean_data %>% 
  select(-c(OULR_locus1841_280,
            OULR_locus10488_266,
            NEO_locus17912_112,
            NEO_locus17285_283,
            OUOU_locus11827_147,
            OUOU_locus6374_33))

# Omit 2 poor performing samples
clean_data <- clean_data %>%
  filter(Sample_ID != "OUACH007") %>%
  filter(Sample_ID != "OUACH008")
```

<b>After filtering, the data set consists of:</b> <br>
<i>N</i><sub>(sample)</sub> = 482 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

#### 1b: Generate metadata; Run the Rmd chunk below to parse metadata from full dataset. 

##### Generate metadata: 
```{r}
# Separate out full metadata only
metadata <- clean_data[,c(1:13)]

# Save full metadata as .Rda file to be loaded in downstream analysis
save(metadata, file = "data/raw_data/snolh_metadata.rda")
```

##### 1b.1. The metadata is saved as the .Rda file `data/raw_data/snolh_all_metadata.rda`, which can be loaded independently for downstream analyses. The associated R object is called "metadata"

#### 1c: Convert SNP genotype data to genepop compatible format; run the Rmd chunk below:
Genepop data consists of genotypes in six digit format, where each allele is a three digit number. Here, we convert SNP nucleotides into two three-digit alleles arbitrarily as follows):

<b>SNP conversion codes:</b> <br>
<b>C</b>: 100 <br>
<b>G</b>: 110 <br>
<b>T</b>: 120 <br>
<b>A</b>: 130 <br>
<b>Missing genotypes</b>: 000 <br>
<b>":"</b> is dropped <br>

e.g., a heterozygous genotype of "A:C" converts to "130100"; a homozygous genotpye of "T:T" converts to "120120"

##### Convert and generate SNP genotype data:
```{r}
# Separate out full genotype data only - do this to retain the nucleotide information if needed!
genotype_data <- clean_data[,c(14:199)]

## Convert SNPs to genepop compatible format
genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("C", "100", x) })) 

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("G", "110", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("T", "120", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("A", "130", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("uknown", "000000", x) })) ## Missing data are demarcated as '000000'

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub(":", "", x) }))

# Save SNP genotype data as .Rda file to be loaded in downstream analysis
save(genotype_data, file = "data/raw_data/snolh_genotype_data.rda")
```

##### 1c.1. The SNP genotype data is saved as the .Rda file `data/raw_data/snolh_genotype_data.rda`, which can be loaded independently for downstream analyses. The associated R object is called "genotype_data".

#### 1d: Join cleaned metadata and converted genotype data into a single full dataset and save for downstream analyses.

##### Join metadata and genotype data into full dataset
```{r}
# cbind metadata and genotype data
full_data <- cbind(metadata, genotype_data)

# save the full dataset for future analyses
save(full_data, file = "data/raw_data/snolh_all_data.rda")
```

##### 1a.1. The cleaned full dataset is saved as the .Rda file `data/raw_data/snolh_all_data.rda`, which can be loaded independently for downstream analyses. The associated R object is called "full_data".

## ------------------------ END OF PHASE 1 ----------------------- ##

<<<<<<< HEAD
## PHASE 2: STRUCTURE ANALYSIS

Programs need:

STRUCTURE v.2.3.4. (Pritchard et al. 2000)

Citation:

Pritchard JK, Stephens M, Donnelly P. 2000. Inference of population structure using multilocus genotype data. Genetics 155: 945-959.

CLUMPP v.1.1.2 (Jakobsson and Rosenberg 2007)

Citation:

Jakobsson M, Rosenberg NA. 2007. CLUMPP: A cluster matching and permutation program for dealing with label switching and multimodality in analysis of population structure. Bioinformatics 23: 1801-1806.

### STEP 1: Conduct population structure and hybridization analysis of Spotted Bass and all other Interior Highlands species
In this step, we are assessing population structure and hybridization between Spotted Bass and all other Interior Highlands species, including Smallmouth Bass, Neosho Bass, Ouachita Bass, and Little River Bass.

#### 1a: Run the Rmd file: `01_snolh_structure_spb_other.Rmd`















### STEP 2: Conduct population structure and hybridization analysis of Smallmouth Bass and all other Interior Highlands species
In this step, we are assessing population structure and hybridization between Smallmouth Bass and all other Interior Highlands species, including Neosho Bass, Ouachita Bass, and Little River Bass.


### ---------------- LOAD PROCESSED DATA HERE ---------------- ### 

#### 3b: Load processed genotype data for downstream analyses; Run the Rmd chunk below:

##### Load data:
```{r}
load("data/processed_genotype_data/spb_other.rda")
load("data/processed_genotype_data/smb_ih.rda")
load("data/processed_genotype_data/ih.rda")
load("data/processed_genotype_data/ouachita_little.rda")
load("data/processed_genotype_data/ouachita.rda")
load("data/processed_genotype_data/little.rda")
```


#### 1c: Compress each output directory into a zip file compatible with Structure Selector (Li and Liu 2017) or Structure Harvester (Earl and vonHoldt 2011) online.

### STEP 2: Submit each zip directory to Structure Selector or Structure Harvester to extract summary results

#### 2a: Copy detlaK tables (Evanno et al. YEAR) for each data combination and save as excel files here: `data/structure_data/summary_data`

#### 2b: Copy puechmaille tables (Puechmaille et al. 2016.) for each data combination and save as excel files here:
`data/structure_data/summary_data`

### STEP 3: Visualize STRUCTURE runs for all data combinations

#### 3a: Convert STRUCTURE files from each analysis into aligned Q files compatible with analysis in the program CLUMPP (Jakobbson and Rosenberg 2007); run the rmd chunk below.

##### Convert STRUCTURE files to aligned Q files for CLUMPP: 
```{r}
# Get a list of structure files for each data combination
spb_other_sfiles <- list.files("data/structure_data/output_data/spb_other_structure_output/", full.names = T)
smb_ih_sfiles <- list.files("data/structure_data/output_data/smb_ih_structure_output/", full.names = T)
ih_sfiles <- list.files("data/structure_data/output_data/ih_structure_output/", full.names = T)
ouachita_little_sfiles <- list.files("data/structure_data/output_data/ouachita_little_structure_output/", full.names = T)
ouachita_sfiles <- list.files("data/structure_data/output_data/ouachita_structure_output/", full.names = T)
little_sfiles <- list.files("data/structure_data/output_data/little_structure_output/", full.names = T)

# Extract q value information (ancestry proportions) from each run for each K value for each individual from the STRUCTURE output files in the directories listed above
spb_other_Q <- readQ(spb_other_sfiles)
smb_ih_Q <- readQ(smb_ih_sfiles)
ih_Q <- readQ(ih_sfiles)
ouachita_little_Q <- readQ(ouachita_little_sfiles)
ouachita_Q <- readQ(ouachita_sfiles)
little_Q <- readQ(little_sfiles)

# Tabulate information from the q lists
sbp_other_tab <- tabulateQ(spb_other_Q)
smb_ih_tab <- tabulateQ(smb_ih_Q)
ih_tab <- tabulateQ(ih_Q)
ouachita_little_tab <- tabulateQ(ouachita_little_Q)
ouachita_tab <- tabulateQ(ouachita_Q)
little_tab <- tabulateQ(little_Q)

# Summarize information from tabultions above
spb_other_summary <- summariseQ(sbp_other_tab)
smb_ih_summary <- summariseQ(smb_ih_tab)
ih_summary <- summariseQ(ih_tab)
ouachita_little_summary <- summariseQ(ouachita_little_tab)
ouachita_summary <- summariseQ(ouachita_tab)
little_summary <- summariseQ(little_tab)

# Extract deltaK and associated summary information using Evanno method
spb_other_evanno <- evannoMethodStructure(spb_other_summary, returnplot = F)
smb_ih_evanno <- evannoMethodStructure(smb_ih_summary, returnplot = F) 
ih_evanno <- evannoMethodStructure(ih_summary, returnplot = F) 
ouachita_little_evanno <- evannoMethodStructure(ouachita_little_summary, returnplot = F) 
ouachita_evanno <- evannoMethodStructure(ouachita_summary, returnplot = F) 
little_evanno <- evannoMethodStructure(little_summary, returnplot = F) 

# Align replicate runs for each K to correct label switching
spb_other_align <- alignK(spb_other_Q)
smb_ih_align <- alignK(smb_ih_Q)
ih_align <- alignK(ih_Q)
ouachita_little_align <- alignK(ouachita_little_Q)
ouachita_align <- alignK(ouachita_Q)
little_align <- alignK(little_Q)
```

#### 3a: Export CLUMPP compatbile files for CLUMPP analysis; run the rmd chunk below to export CLUMPP associated files for each K for each analysis.

<b>For each analysis, we used the following parameters: </b> <br>

Large-K-Greedy algorithm (paramrep = 3) <br>
10000 replicates <br>

### IMPORTANT NOTE: This step only needs to be run ONCE to generate files for CLUMPP. Once you have run this chunk, move on to Step 3b. Uncomment each line to run this code

##### Export CLUMPP files:
```{r}
# clumppExport(spb_other_align, 
            # parammode = 3, 
            # paramrep = 10000,
            # exportpath = "data/structure_data/clumpp_data/spb_other_clumpp")
 
# clumppExport(smb_ih_align, 
            # parammode = 3, 
            # paramrep = 10000, 
            # exportpath = "data/structure_data/clumpp_data/smb_ih_clumpp")

# clumppExport(ih_align, 
            # parammode = 3,
            # paramrep = 10000, 
            # exportpath = "data/structure_data/clumpp_data/ih_clumpp")

# clumppExport(ouachita_little_align, 
            # parammode = 3,
            # paramrep = 10000, 
            # exportpath = "data/structure_data/clumpp_data/ouachita_little_clumpp")

# clumppExport(ouachita_align, 
            # parammode = 3,
            # paramrep = 10000, 
            # exportpath = "data/structure_data/clumpp_data/ouachita_clumpp")

# clumppExport(little_align, 
            # parammode = 3,
            # paramrep = 10000, 
            # exportpath = "data/structure_data/clumpp_data/little_clumpp")
```

The code above generates a .txt file (pop_K#-combined.txt) with combined cluster ancestry proportions for each individual at each K and stores it in a separate directory labeled "pop_K#", where '#' is the corresponding K value. Additionally, the code generates an accompanying "paramfile" that is input for CLUMPP, which contains information on parameters to generate merged datasets across our 10 replicates at each K.

#### 3b: Generate merged Q files using CLUMPP.
In this step, we used the software program CLUMPP (Jakobbson and Rosenberg 2007) to merge cluster ancestry proportion inferences for all replicate runs of each K across individuals. We downloaded the Linux 64-bit build of CLUMPP v.1.1.2 and installed it on our computing cluster. We then executed the program for each analysis by sequentially copying the associated paramfiles generated in the clumppExport function (see section above) to the CLUMPP home directory. All output files were then moved to the corresponding structure directory.

##### 3b.1. Downolad CLUMPP and install on cluster; navigate to the desired directory: `code`. Run the command: `wget https://rosenberglab.stanford.edu/software/CLUMPP_Linux64.1.1.2.tar.gz`

###### 3b.1.1. Run the command: `gunzip CLUMPP_Linux64.1.1.2.tar.gz`

###### 3b.1.2. Run the command: `tar -xvf CLUMPP_Linux64.1.1.2.tar`

##### 3b.2. For each analysis separately, copy the 'pop_K#-combined.txt' file and 'paramfile' over to the CLUMPP directory (`code/CLUMPP_Linux64.1.1.2`)

##### 3b.3. For each analysis separately, execute CLUMPP on the paramfile; run the code `./CLUMPP paramfile`
We are running this code with GREEDY OPTION 2, which uses random sampling of replicate runs. 

This code generates three additional files, which we moved back to the corresponding pop_K folder in the clumpp_data directory:

<b>Files generated: </b> <br>

pop_K#-combined-miscfile.txt <br>
pop_K#-combined-merged.txt <br>
pop_K#-combined-aligned.txt <br>

#### 3c: Plot STRUCTURE runs at best K value (according to deltak statistic) for each analysis.

##### 3c.1. Plot results for SPB AND ALL OTHER INTERIOR HIGHLANDS POPULATIONS; run the rmd chunk below.

##### STRUCTURE results of SPB and all other Interior Highlands populations
```{r}
# DeltaK results
ggplot(spb_other_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 2, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(spb_other_evanno$k), breaks = spb_other_evanno$k)

# convert metadata columns to characters 
spb_other <- spb_other %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

spb_other_k2 <- readQ("data/structure_data/clumpp_data/spb_other_clumpp/pop_K2/pop_K2-combined-merged.txt")

# Visualize Structure
plotQ(spb_other_k2,
      grplab = spb_other[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("mediumpurple", "springgreen3"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "spb_other_k2",
      exportpath = "figures/structure_figures")
```

##### 3c.2. Plot results for SMB AND ALL OTHER INTERIOR HIGHLANDS POPULATIONS; run the rmd chunk below.

##### STRUCTURE results of SMB and all other Interior Highlands populations
```{r}
# DeltaK results
ggplot(smb_ih_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 2, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(smb_ih_evanno$k), breaks = smb_ih_evanno$k)

# convert metadata columns to characters 
smb_ih <- smb_ih %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

smb_ih_k2 <- readQ("data/structure_data/clumpp_data/smb_ih_clumpp/pop_K2/pop_K2-combined-merged.txt")

# Visualize Structure
plotQ(smb_ih_k2,
      grplab = smb_ih[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("springgreen3", "firebrick3"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "smb_ih_k2",
      exportpath = "figures/structure_figures")
```

##### 3c.3. Plot results for INTERIOR HIGHLANDS POPULATIONS; run the rmd chunk below.

##### STRUCTURE results for Interior Highlands populations
```{r}
# DeltaK results
ggplot(ih_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 3, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(ih_evanno$k), breaks = ih_evanno$k)

# convert metadata columns to characters 
ih <- ih %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

ih_k3 <- readQ("data/structure_data/clumpp_data/ih_clumpp/pop_K3/pop_K3-combined-merged.txt")

# Visualize Structure
plotQ(ih_k3,
      grplab = ih[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("springgreen3", "goldenrod3","deepskyblue"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "ih_k3",
      exportpath = "figures/structure_figures")
```

##### 3c.4. Plot results for OUACHITA AND LITTLE RIVER BASS only; run the rmd chunk below.

##### STRUCTURE results for Ouachita and Little River populations
```{r}
# DeltaK results
ggplot(ouachita_little_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 2, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(ouachita_little_evanno$k), breaks = ouachita_little_evanno$k)

# convert metadata columns to characters 
ouachita_little <- ouachita_little %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

ouachita_little_k2 <- readQ("data/structure_data/clumpp_data/ouachita_little_clumpp/pop_K2/pop_K2-combined-merged.txt")

#Visualize Structure
plotQ(ouachita_little_k2,
      grplab = ouachita_little[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("goldenrod3","deepskyblue"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "ouachita_little_k2",
      exportpath = "figures/structure_figures")
```

##### 3c.5. Plot results for OUACHITA BASS only; run the rmd chunk below.

##### STRUCTURE results for Ouachita populations
```{r}
# DeltaK results
ggplot(ouachita_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 2, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(ouachita_evanno$k), breaks = ouachita_evanno$k)

# convert metadata columns to characters 
ouachita <- ouachita %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

ouachita_k2 <- readQ("data/structure_data/clumpp_data/ouachita_clumpp/pop_K2/pop_K2-combined-merged.txt")

# Visualize Structure
plotQ(ouachita_k2,
      grplab = ouachita[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("deepskyblue","deeppink2"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "ouachita_k2",
      exportpath = "figures/structure_figures")
```

##### 3c.6. Plot results for LITTLE RIVER BASS only; run the rmd chunk below.

##### STRUCTURE results for Little River populations
```{r}
# DeltaK results
ggplot(little_evanno, aes(x=k, y=deltaK)) +
  geom_point() + 
  geom_line() +
  theme_cowplot(theme_set(12)) +
  geom_vline(xintercept = 3, color = "blue") +
  theme(axis.text = element_text(size = 15)) +
  theme(axis.title = element_text(size = 15)) +
  theme(axis.title.x = element_text(face = "italic")) +
  labs(x = "K", y = "delta K") + 
  scale_x_continuous("K", labels = as.character(little_evanno$k), breaks = little_evanno$k)

# convert metadata columns to characters 
little <- little %>%
  mutate(Structure_Num = as.character(Structure_Num),
         River = as.character(River))

little_k3 <- readQ("data/structure_data/clumpp_data/little_clumpp/pop_K3/pop_K3-combined-merged.txt")

# Visualize Structure
plotQ(little_k3,
      grplab = little[,c(12,6)],
      showindlab = FALSE, 
      ordergrp = F,
      selgrp = "River",
      showgrplab = T,
      showlegend = F, 
      showsp = F, 
      showdiv = T, 
      divsize = 2, 
      divtype = 1,
      divcol = "black",  
      grplabsize = 4, 
      grplabangle = 45,
      grplabjust = 0.1,
      legendkeysize = 15, 
      linealpha = 0,
      legendtextsize = 10, 
      linesize = 0.05, 
      pointsize = 4, 
      barbordercolour = "black",
      barbordersize = 0.1, 
      clustercol = c("goldenrod3","sienna4","lightyellow"),
      exportplot = T,
      imgtype = "pdf",
      height = 10, 
      width = 100,
      outputfilename = "little_k3",
      exportpath = "figures/structure_figures")
```

=======
## PHASE 2: STRUCTURE AND HYBRID ASSIGNMENT

<b>Programs need</b>: <br>

STRUCTURE v.2.3.4. (Pritchard et al. 2000) <br>

<b>Citation</b>: <br>

Pritchard JK, Stephens M, Donnelly P. 2000. Inference of population structure using multilocus genotype data. Genetics 155: 945-959. <br><br>


CLUMPP v.1.1.2 (Jakobsson and Rosenberg 2007) <br>

<b>Citation</b>: <br>

Jakobsson M, Rosenberg NA. 2007. CLUMPP: A cluster matching and permutation program for dealing with label switching and multimodality in analysis of population structure. Bioinformatics 23: 1801-1806. <br><br>


NEWHYBRIDS v. 2.0 (Anderson) <br>

<b>Citation</b>: <br>

Anderson EC, Thompson EA. 2002. A model-based method for identifying species hybrids using multilocus genetic data. Genetics 160: 1217-1229. <br><br>


PARALLELNEWHYBRID v.0.0.0.9002

<b>Citation</b>: <br>

Wringe BF, Stanley RRE, Jeffery NW, Anderson EC, Bradbury IR. 2017. <i>parallelnewhybrid</i>: An R package for the parallelization of hybrid detection using NEWHYBRIDS. Molecular Ecology Resources 17: 91-95


### STEP 1: Conduct population structure and hybridization analysis 

#### 1a: Assess Spotted Bass and Interior Highlands species; run the Rmd file: `01_snolh_structure_spb_other.Rmd`
In this step, we are assessing population structure and hybridization between Spotted Bass and all other Interior Highlands species, including Smallmouth Bass, Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Spotted Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross). We then omit any detected hybrids from the dataset and continue with population structure and hybridization analysis between Smallmouth Bass and the rest of the Interior Highland species in a hierarchical fashion.

#### 1b: Assess Smallmouth Bass and Interior Highlands species; run the Rmd file: `02_snolh_structure_smb_ih.Rmd`
In this step, we are assessing population structure and hybridization between Smallmouth Bass and all other Interior Highlands species, including Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Smallmouth Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross). We then omit any detected hybrids from the dataset and continue with population structure and hybridization analysis among only the rest of the Interior Highland species in a hierarchical fashion.

#### 1b: Assess Interior Highlands species only; run the Rmd file: `03_snolh_structure_ih.Rmd`
In this step, we are assessing population structure and hybridization between all  Interior Highlands species only, including Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Smallmouth Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross).
>>>>>>> 01b1ed48716f80e224cfd81f79a4944bd3f2a21c
