---
title: "Analysis 1: Structure and Hybrid Assignment Analysis"
author: "Joe Gunn"
date: "2022-07-28"
output: html_document
---

# Project: Interior Highlands Hybridization: Spotted, Smallmouth, Neosho, Ouachita, and Little River Bass (SNOLH)
<font size="+1">Investigating hybridization and population structure among and within the Spotted Bass (<i>Micropterus punctulatus</i>), Smallmouth Bass (<i>M. dolomieu</i>), the newly elevated Neosho Bass (<i>M. velox</i>), and two other potentially distinct species in the Ouachita River Basin (the Ouachita and Little River Basses)</font>

## Specific Aim: Hierarchical structure analysis with diagnostic markers
For this aim, we are using Bayesian clustering analysis with STRUCTURE to assess patterns and levels of hybridization among and within black bass species in the Interior Highlands, with emphasis on populations in the Arkansas River Basin and Ouachita River Basin. We obtained SNP marker data from from Spotted Bass (<i>M. punctulatus</i>), Smallmouth Bass (<i>M. dolomieu</i>), Neosho Bass (<i>M. velox</i>), and two potentially distinct species in the Ouachita River Basin (Ouachita Bass and Little River Bass) which were known to be diagnostic of these species (Long et al. 2021).

## Phases of Analysis
### Phase 1: Data preparation
### Phase 2: Data summarization
### Phase 3: Structure and hybrid assignment analysis

### Libraries needed for analysis
```{r}
library(tidyverse)
library(cowplot)
library(readxl)
library(writexl)
library(genepopedit)
library(pophelper)
```

## PHASE 1: DATA PREPARATION 

### STEP 1: Read in full dataset and generate metadata
In this step, we are reading in the raw data file (which includes sample metadata and SNP genotype data; 'snolh_all_data.xlsx'), manipulating and cleaning the data, and generating separate data frames for the metadata (without SNP genotype data) and for the SNP genotype data (without metadata). We will then be able to break the metadata and SNP data down into separate subsets of species/populations/markers, etc., and these subsets can be pieced together for subsequent hierarchical analyses.

### ***Skip to Step 1d to load in raw data objects immediately***
Steps 1a - 1c should be initially run to generate saved .Rda files. Once these files have been saved, there is no need to run Steps 1a - 1c again. We can skip directly to Step 1d to load in the data.

### ***Skip to Step 3b to load in processed genotype data objects (for structure analysis) immediately***
Steps 2a - 3a should be initially run after loading in raw data objects (see above) to generate saved .Rda files. Once these files have been saved, there is no need to run Steps 2a - 3a again. We can skip directly to Step 3b to load in the data.

#### 1a: Read in, clean, and save full dataset. 

<b>Before filtering:</b> <br>
<i>N</i><sub>(sample)</sub> = 487 <br>
<i>N</i><sub>(loci)</sub> = 192 <br>

##### IMPORTANT NOTE ON DATA CLEANING: As part of data cleaning, we are omitting three samples representing Shoal Bass (<i>M. cataractae</i>) due to low sample size and because we are not explicitly interested in Shoal Bass hybridization or population structure. Additionally, we remove SNP markers with low call rates across samples and samples with low genotyping rates across SNPs.

##### 1a.1. Read in full dataset, convert characters to factors, and omit Shoal Bass samples; run the Rmd chunk below.

##### Read in, clean, and save full dataset: `../raw_data/snolh_all_data.xlsx`
```{r}
# Read in raw dataset, with metadata appended
raw_data <- read_excel("../raw_data/snolh_all_data.xlsx") 

# Convert characters to factors
clean_data <- raw_data %>%
  mutate(Ref_samps = factor(Ref_samps), 
         Sample_Num = factor(Sample_Num), 
         Putative_Taxon = factor(Putative_Taxon),
         Vis_ID = factor(Vis_ID),
         River = factor(River),
         Location = factor(Location),
         Date = factor(Date),
         Genetics_Num = factor(Genetics_Num),
         Structure_Num = factor(Structure_Num),
         Sample_ID = factor(Sample_ID))

# Omit the 3 shoal bass samples in the first 3 rows.
clean_data <- clean_data[-c(1:3),] 
```

<b>Filtering results:</b> <br>
<i>N</i><sub>(sample)</sub> = 484 <br>
<i>N</i><sub>(loci)</sub> = 192 <br>

##### 1a.2. Filter the dataset for poor quality SNP loci; run the Rmd chunk below.
Here, we omit any SNP loci from the dataset with over 20% (rate of 0.20) missing data across samples. 

#### Filter SNPs with low genotype call rates across samples:
```{r}
# Get counts of "no-genotype" calls per site across samples
snp_missing <- apply(clean_data[14:205], 2, function(x) {length(which(x=="unknown"))})

# Convert to dataframe
snp_missing <- as.data.frame(snp_missing)

# Convert rownames (which are SNP Ids) to column and call it "snp_id"
snp_missing <- rownames_to_column(snp_missing, "snp_id")

# Calculate rate of missing genotypes per locus (number of missing genotypes divided by total number of samples (484 after shoal bass filtering))
snp_missing <- snp_missing %>%
  mutate(missing_rate = snp_missing/484)

# View genotyping missingness in descending order for filtering 
snp_missing %>%
  arrange(desc(missing_rate))

# Omit 6 poor performing loci with over 20% missing data
clean_data <- clean_data %>% 
  select(-c(OULR_locus1841_280,
            OULR_locus10488_266,
            NEO_locus17912_112,
            NEO_locus17285_283,
            OUOU_locus11827_147,
            OUOU_locus6374_33))
```
<b>Filtering results:</b> <br>

<b>The following LOCI were removed from further analysis:</b> <br>
OULR_locus1841_280 <br>
NEO_locus17912_112 <br>
OULR_locus10488_266 <br>
OUOU_locus11827_147 <br>
NEO_locus17285_283 <br>
OUOU_locus6374_33 <br>

<i>N</i><sub>(sample)</sub> = 484 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

##### 1a.3. Filter the dataset for poor quality samples; run the Rmd chunk below.
Here, we omit any samples from the dataset with over 20% (rate of 0.20) missing data across SNP loci 

#### Filter samples with low genotype call rates across SNP loci:
```{r}
# Get counts of "no-genotype" calls per sample across loci
sample_missing <- apply(clean_data, 1, function(x) {length(which(x=="unknown"))})

# Convert to dataframe 
sample_missing <- as.data.frame(sample_missing)

# Bind with sample ID
sample_missing <- cbind(clean_data$Sample_ID, sample_missing)

# Change first column name
colnames(sample_missing) <- c("sample_id", "sample_missing")

# Calculate rate of missing genotypes per sample (number of missing genotypes divided by total number of SNP loci (186 after poor quality SNP filtering filtering))
sample_missing <- sample_missing %>%
  mutate(missing_rate = sample_missing/186)

# View genotyping missingness in descending order for filtering 
sample_missing %>%
  arrange(desc(missing_rate))

# Omit 2 poor performing samples
clean_data <- clean_data %>%
  filter(Sample_ID != "OUACH007") %>%
  filter(Sample_ID != "OUACH008") %>%
  filter(Sample_ID != "G1018-03-OSU_SKIA-4") %>%
  filter(Sample_ID != "G1018-03-OSU_GRSPB-001") %>%
  filter(Sample_ID != "G1018-03-OSU_SPVW-004") %>%
  filter(Sample_ID != "OUACH015")
```

<b>The following SAMPLES were removed from further analysis:</b> <br>
OUACH007 <br>
OUACH008 <br>
G1018-03-OSU_SKIA-4 <br>
G1018-03-OSU_GRSPB-001 <br>
G1018-03-OSU_SPVW-004 <br>
OUACH015 <br>

<b>After filtering, the data set consists of:</b> <br>
<i>N</i><sub>(sample)</sub> = 478 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

##### 1a.4. Filter the dataset for potential (likely) duplicate samples based on percent identity.
The Center of Aquaculture Technologies (CAT), who genotyped our black bass samples, used percent identity to detect potential duplicate samples in the dataset. They identified 14 samples as potential duplicates with a percent identity over 95% (0.95). Two of those samples represented Shoal Bass and were already removed in step 1a.1 above. We removed the remaining 12 samples from all downstream analyses.

#### Filter duplicate samples:
```{r}
# Omit 2 poor performing samples
clean_data <- clean_data %>%
  filter(Sample_ID != "GLVR-011") %>%
  filter(Sample_ID != "UMF_045") %>%
  filter(Sample_ID != "G1018-03-OSU_GLVR-006") %>%
  filter(Sample_ID != "UMF006") %>%
  filter(Sample_ID != "G1018-03-OSU_BFC-023") %>%
  filter(Sample_ID != "G1018-03-OSU_GLVR-005") %>%
  filter(Sample_ID != "GLVR-022") %>%
  filter(Sample_ID != "UMF035") %>%
  filter(Sample_ID != "GLVR-008") %>%
  filter(Sample_ID != "UMF012") %>%
  filter(Sample_ID != "BFC061") %>%
  filter(Sample_ID != "G1018-03-OSU_GLVR-024")
```
<b>The following SAMPLES were removed from further analysis:</b> <br>
GLVR-011 <br>
UMF_045 <br>
G1018-03-OSU_GLVR-006 <br>
UMF006 <br>
G1018-03-OSU_BFC-023 <br>
G1018-03-OSU_GLVR-005 <br>
GLVR-022 <br>
UMF035 <br> 
GLVR-008 <br> 
UMF012<br> 
BFC061 <br> 
G1018-03-OSU_GLVR-024 <br>

<b>After filtering, the data set consists of:</b> <br>
<i>N</i><sub>(sample)</sub> = 466 <br>
<i>N</i><sub>(loci)</sub> = 186 <br>

##### 1a.5. Save the fully filtered, clean dataset for future analyses; run the Rmd chunk below

##### Save fully filtered, clean dataset: `../raw_data/snolh_clean_data.Rda` 
```{r}
# Save full, clean dataset as .Rda file to be loaded in downstream analysis
save(clean_data, file = "../raw_data/snolh_clean_data.rda")
```

#### 1b: Generate metadata; Run the Rmd chunk below to parse metadata from full dataset. 

##### Generate metadata: 
```{r}
# Separate out full metadata only
metadata <- clean_data[,c(1:13)]

# Save full metadata as .Rda file to be loaded in downstream analysis
save(metadata, file = "../raw_data/snolh_metadata.rda")
```

##### 1b.1. The metadata is saved as the .Rda file `data/raw_data/snolh_all_metadata.rda`, which can be loaded independently for downstream analyses. The associated R object is called "metadata"

#### 1c: Convert SNP genotype data to genepop compatible format; run the Rmd chunk below:
Genepop data consists of genotypes in six digit format, where each allele is a three digit number. Here, we convert SNP nucleotides into two three-digit alleles arbitrarily as follows):

<b>SNP conversion codes:</b> <br>
<b>C</b>: 100 <br>
<b>G</b>: 110 <br>
<b>T</b>: 120 <br>
<b>A</b>: 130 <br>
<b>Missing genotypes</b>: 000 <br>
<b>":"</b> is dropped <br>

e.g., a heterozygous genotype of "A:C" converts to "130100"; a homozygous genotpye of "T:T" converts to "120120"

##### Convert and generate SNP genotype data:
```{r}
# Separate out full genotype data only - do this to retain the nucleotide information if needed!
genotype_data <- clean_data[,c(14:199)]

## Convert SNPs to genepop compatible format
genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("C", "100", x) })) 

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("G", "110", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("T", "120", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("A", "130", x) }))

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub("uknown", "000000", x) })) ## Missing data are demarcated as '000000'

genotype_data <- data.frame(lapply(genotype_data, function(x) {
  gsub(":", "", x) }))

# Save SNP genotype data as .Rda file to be loaded in downstream analysis
save(genotype_data, file = "data/raw_data/snolh_genotype_data.rda")
```

##### 1c.1. The SNP genotype data is saved as the .Rda file `data/raw_data/snolh_genotype_data.rda`, which can be loaded independently for downstream analyses. The associated R object is called "genotype_data".

#### 1d: Join cleaned metadata and converted genotype data into a single full dataset and save for downstream analyses.

##### Join metadata and genotype data into full dataset
```{r}
# cbind metadata and genotype data
full_data <- cbind(metadata, genotype_data)

# save the full dataset for future analyses
save(full_data, file = "../raw_data/snolh_all_data.rda")
```

##### 1d.1. The cleaned full dataset is saved as the .Rda file `data/raw_data/snolh_all_data.rda`, which can be loaded independently for downstream analyses. The associated R object is called "full_data".

## ------------------------ END OF PHASE 1: DATA PREPARATION  ----------------------- ##

## PHASE 2: DATA SUMMARIZATION
In this phase of the analysis, we are summarizing the full dataset to determine the total number of samples, the number of sites covered, the number of samples per species, etc., for inclusion in the final manuscript.

### STEP 1: Sample summary counts
In this step, we are counting the total number of samples (before and after filtering; see Step 1a above) across sample sites and obtaining counts of samples per species.

#### 1a: Count samples in the full dataset; run the Rmd chunk below:

##### Sample count summaries:
```{r}
## Get counts of individuals per species
full_data %>%
  group_by(Putative_Taxon) %>%
  count()

## Get counts of individuals per river per species
full_data %>%
  group_by(Putative_Taxon, River) %>%
  count()
```

## PHASE 3: STRUCTURE AND HYBRID ASSIGNMENT ANALYSIS

Programs need:

STRUCTURE v.2.3.4. (Pritchard et al. 2000)

Citation:

Pritchard JK, Stephens M, Donnelly P. 2000. Inference of population structure using multilocus genotype data. Genetics 155: 945-959.

CLUMPP v.1.1.2 (Jakobsson and Rosenberg 2007)

Citation:

Jakobsson M, Rosenberg NA. 2007. CLUMPP: A cluster matching and permutation program for dealing with label switching and multimodality in analysis of population structure. Bioinformatics 23: 1801-1806.

### STEP 1: Conduct population structure and hybridization analysis 

#### 1a: Assess Spotted Bass and Interior Highlands species; run the Rmd file: `01_snolh_structure_spb_other.Rmd`
In this step, we are assessing population structure and hybridization between Spotted Bass and all other Interior Highlands species, including Smallmouth Bass, Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Spotted Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross). We then omit any detected hybrids from the dataset and continue with population structure and hybridization analysis between Smallmouth Bass and the rest of the Interior Highland species in a hierarchical fashion.

#### 1b: Assess Smallmouth Bass and Interior Highlands species; run the Rmd file: `02_snolh_structure_smb_ih.Rmd`
In this step, we are assessing population structure and hybridization between Smallmouth Bass and all other Interior Highlands species, including Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Smallmouth Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross). We then omit any detected hybrids from the dataset and continue with population structure and hybridization analysis among only the rest of the Interior Highland species in a hierarchical fashion.

#### 1c: Assess Interior Highlands species only; run the Rmd file: `03_snolh_structure_ih.Rmd`
In this step, we are assessing population structure and hybridization between all  Interior Highlands species only, including Neosho Bass, Ouachita Bass, and Little River Bass. We use the program STRUCTURE to estimate ancestry coefficients and assess overall population structure among stream populations within the Smallmouth Bass and Interior Highlands ranges, and then we use NEWHYBRIDS analysis to assign any putative hybrids to one of six hybrid categories (pure parent 1, pure parent 2, F1 hybrid, F2 hybrid, parent 1 backcross, or parenet 2 backcross).
